---
title: "Omicron_ACE2_Landscape"
author: "Alief Moulana"
date: "8/27/2022"
output: rmarkdown::github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(plyr)
library(ggplot2)
library(RColorBrewer)
library(stringr)
library(gtools)
library(ggpmisc)
library(tidyr)
library(dplyr)
library(ggnewscale)
library(ggrepel)
library(paletteer)
library(cowplot)
library(Rmisc)
library(scales)
library(tidyverse)
library(ggtext)
library(purrr)
```

## Load the Data
Before running any chunk to generate figures, load all required datasets. This is to ensure you have downloaded all the datasets in the `data` category.
```{r Data_Loading}
###Mutation Dictionary###
n <- 1:15 #initialize 15 mutations
nn <- c(339, 371, 373, 375, 417, 440, 446, 477, 478, 484, 493, 496, 498, 501, 505) #initialize the correct numbering for the mutations
nn.muts <- c("G339D","S371L","S373P","S375F","K417N","N440K","G446S","S477N","T478K","E484A","Q493R","G496S","Q498R","N501Y","Y505H")
AppendMe <- function(dfNames) {
  do.call(rbind, lapply(dfNames, function(x) {
    cbind(get(x), Antibody = x)
  }))
}
###KD Data###
#initialize, get Kd data

CoV555 <- read.table(file = 'data/cleaned_Kds_RBD_CoV555_proper.csv', sep = ',', 
                  header = TRUE,
                  colClasses=c("numeric","character",rep("numeric",57),"logical","logical")
                  ) %>% select(geno,log10Kd_pinned,err_log10Kd,is_non_binder,is_clean)
Kds.CoV555 <- CoV555  

S309 <- read.table(file = 'data/cleaned_Kds_RBD_S309_proper.csv', sep = ',', 
                  header = TRUE,
                  colClasses=c("numeric","character",rep("numeric",65),"logical","logical")
                  ) %>% select(geno,log10Kd_pinned,err_log10Kd,is_non_binder,is_clean)
Kds.S309 <- S309  

CB6 <- read.table(file = 'data/cleaned_Kds_RBD_CB6_proper.csv', sep = ',', 
                  header = TRUE,
                  colClasses=c("numeric","character",rep("numeric",61),"logical","logical")
                  ) %>% select(geno,log10Kd_pinned,err_log10Kd,is_non_binder,is_clean)
Kds.CB6 <- CB6  

REGN10987 <- read.table(file = 'data/cleaned_Kds_RBD_REGN10987_proper.csv', sep = ',', 
                  header = TRUE,
                  colClasses=c("numeric","character",rep("numeric",63),"logical","logical")
                  ) %>% select(geno,log10Kd_pinned,err_log10Kd,is_non_binder,is_clean)
Kds.REGN10987 <- REGN10987  

Kds_ACE2 <- read.table(file = 'data/cleaned_Kds_RBD_ACE2_low_filter.tsv', sep = '\t', 
                  header = TRUE, 
                  colClasses=c("character",rep("numeric",45))
                  ) %>% 
  select(geno,log10Kd_pinned,err_log10Kd) %>%
  mutate(Antibody = "ACE2")
ACE2 <- Kds_ACE2[,1:3]
ACE2$is_non_binder <- FALSE
ACE2$is_clean <- TRUE
Kds.ACE2 <- ACE2

Kds.all <- AppendMe(c("CoV555","S309","CB6","REGN10987","ACE2"))
Kds_simplified <- data.frame(Genotype=Kds.all$geno,
                             Log10Kd_unfiltered = Kds.all$log10Kd_pinned, 
                             Error = Kds.all$err_log10Kd,
                             Antibody = Kds.all$Antibody,
                             NonBinder = Kds.all$is_non_binder,
                             Filtered = Kds.all$is_clean) #drop most of the columns
Kds_simplified$Log10Kd <- ifelse((Kds_simplified$NonBinder),
                                 5.0,
                                 ifelse(Kds_simplified$Filtered,
                                        Kds_simplified$Log10Kd_unfiltered,
                                        NA))
#here what we do is to pin NonBinder to 5.0 (this value is just to make it easier to plot) and if the data is not clean, we also exclude them

Kds.short.tradeoff <- data.frame(Genotype = Kds_simplified[1:nrow(Kds.ACE2),]$Genotype,
                                 ACE2 = subset(Kds_simplified,Antibody=="ACE2")$Log10Kd,
                                 CB6 = subset(Kds_simplified,Antibody=="CB6")$Log10Kd,
                                 CoV555 = subset(Kds_simplified,Antibody=="CoV555")$Log10Kd,
                                 REGN10987 = subset(Kds_simplified,Antibody=="REGN10987")$Log10Kd,
                                 S309 = subset(Kds_simplified,Antibody=="S309")$Log10Kd)

```

## General Data Clean-up
```{r prepping_figure1c}
Kds_simplified$Antibody <- factor(Kds_simplified$Antibody, levels = c("CB6","CoV555","REGN10987","S309","ACE2"))
Kds_simplified[paste0('Mutation ', nn)] <- purrr::map_dfc(n, 
                                                     ~case_when(substr(Kds_simplified$Genotype,.x,.x) == '0' ~ '0',
                                                                substr(Kds_simplified$Genotype,.x,.x) == '1' ~ '1'
                                                     )
) #create columns that correspond to the allele for each mutation
Kds_simplified[paste0('Mutation ', nn)] <- lapply(Kds_simplified[paste0('Mutation ', nn)], factor) #factorize (object type) the mutations
wuhan.Kd <- data.frame(Antibody=c("CoV555","S309","CB6","REGN10987","ACE2"),
                       Kd=Kds_simplified[Kds_simplified$Genotype==
                             paste(replicate(length(n), "0"), collapse =
                                     ""),"Log10Kd"])
omicron.Kd <- data.frame(Antibody=c("CoV555","S309","CB6","REGN10987","ACE2"),
                         Kd=Kds_simplified[Kds_simplified$Genotype==
                             paste(replicate(length(n), "1"), collapse =
                                     ""),"Log10Kd"])
Kds_simplified$Class <- str_count(Kds_simplified$Genotype,"1") #initialize variable "Class", which is #hamming distance from Wuhan
Kds_simplified$Class <- ordered(Kds_simplified$Class, levels = as.character(0:15)) #order em

#getsomepalettes
getYellowtoRed <- colorRampPalette(c('#ffffcc','#ffeda0','#fed976','#feb24c','#fd8d3c','#fc4e2a','#e31a1c','#bd0026','#800026'))
getPalettePairedLong = colorRampPalette(c('#121212','#808080','#a6cee3','#1f78b4','#b2df8a','#33a02c','#fb9a99','#e31a1c','#fdbf6f','#ff7f00','#cab2d6','#6a3d9a','#ffff99','#b15928','#4f1530'))
binarypermutations <- function(n, vals = 0:1){
  tmp <- rep(list(vals), n)
  var <- do.call(expand.grid, tmp)
  return(apply(var, 1, paste0, collapse=''))
} #create a function that generates all possible 15-tuple permutations between 0 and 1

insertstr <- function(x,y,when){
  a <- x[1]
  b <- y
  return(paste(substr(a,1,when-1),b,substr(a,when,nchar(a)),sep=""))
} #create a function that inserts some character or string into another string in the first column of a dataframe at a specific position
antibodies <- c("CB6","CoV555","REGN10987","S309")

for(i in c(antibodies,"ACE2")){
  Kds.temp <- subset(Kds_simplified,Antibody==i)
  Kds_mutational_effects <- data.frame(Genotype = binarypermutations(14)) #create a dataframe
  Kds_mutational_effects[paste0('Mutation ',nn,"_0")] <- purrr::map_dfc(n, 
                                                            ~with(Kds.temp, Log10Kd[match(
                                                              apply(Kds_mutational_effects,1,insertstr,when=.x,y="0"), Genotype)])
                                                          )  #create a variable of Kd values where at a specific locus, the allele = 0
  Kds_mutational_effects[paste0('Mutation ',nn,"_1")] <- purrr::map_dfc(n, 
                                                            ~with(Kds.temp, Log10Kd[match(
                                                              apply(Kds_mutational_effects,1,insertstr,when=.x,y="1"), Genotype)])
                                                          ) #create a variable of Kd values where at a specific locus, the allele = 1
  Kds_mutational_effects[paste0('Effect_ ',nn)] <- purrr::map_dfc(nn, 
                                                            ~Kds_mutational_effects[paste0('Mutation ',.x,"_1")]-
                                                              Kds_mutational_effects[paste0('Mutation ',.x,"_0")])
  Kds_mutational_effects[paste0('Effect_error_',nn)] <- purrr::map_dfc(n, 
                                                            ~sqrt(
                                                              (with(get(paste0("Kds.",i)), err_log10Kd[match(
                                                              apply(Kds_mutational_effects,1,insertstr,when=.x,y="1"), geno)]))^2+
                                                              (with(get(paste0("Kds.",i)), err_log10Kd[match(
                                                              apply(Kds_mutational_effects,1,insertstr,when=.x,y="0"), geno)]))^2
                                                            )
                                                          )
  #create a set of variables for the difference (mutational effects) and its errors
  Kds_mutational_effects.long <- reshape(Kds_mutational_effects, direction="long", 
          varying=list(paste0('Mutation ',nn,"_0"), paste0('Mutation ',nn,"_1"), paste0('Effect_ ',nn),paste0('Effect_error_',nn)), 
          v.names=c("KD WT","KD OM","KD Effect","Effect Error")) #make the dataframe long. It's not possible to do it above, because of matrix optim
  Kds_mutational_effects.long$Mutation <- as.factor(Kds_mutational_effects.long$time) #create a new variable called mutation, denoting mutational identity we consider in each effect
  levels(Kds_mutational_effects.long$Mutation) <- nn.muts #renaming the levels of the factor
  assign(paste0("Kds_mutational_effects.long.", i),Kds_mutational_effects.long)
  Kx.temp <- get(paste0("Kds_mutational_effects.long.", i))
  Kx.temp$Antibody <- i
  assign(paste0("Kds_mutational_effects.long.", i),Kx.temp)
}


Kds_mutational_effects.long.all.abs <- rbind(Kds_mutational_effects.long.CoV555,
                                     Kds_mutational_effects.long.S309,
                                     Kds_mutational_effects.long.CB6,
                                     Kds_mutational_effects.long.REGN10987,
                                     Kds_mutational_effects.long.ACE2)
data_summary <- function(x) {
   m <- mean(x)
   ymin <- m-sd(x)
   ymax <- m+sd(x)
   return(c(y=m,ymin=ymin,ymax=ymax))
}

Kds_mutation_effects.long.filtered <- subset(Kds_mutational_effects.long.all.abs,
                                             `KD WT` > 6)
```

## Figure 1 - General Landscape
Figure 1A: General Landscape.
```{r histogram}
antibody.colors <- c(getPalettePairedLong(15)[4],getPalettePairedLong(15)[6],
                     getPalettePairedLong(15)[8],getPalettePairedLong(15)[10])
Kds_simplified$Log10Kd.nb <- ifelse(Kds_simplified$Log10Kd < 6.01, 5.15, Kds_simplified$Log10Kd)
omicron.Kd$Kd <- ifelse(omicron.Kd$Kd < 6.01, 5.15, omicron.Kd$Kd)
Kds_simplified.withACE2 <- Kds_simplified
Kds_simplified.withoutACE2 <- subset(Kds_simplified,Antibody %in% antibodies)
Kds_simplified.withoutACE2$Antibody <- factor(Kds_simplified.withoutACE2$Antibody,levels=antibodies)
Kds_simplified.withoutACE2$Antibody.used <- ifelse(Kds_simplified.withoutACE2$Antibody=="CB6","LY-CoV016",
                                             ifelse(Kds_simplified.withoutACE2$Antibody=="CoV555",
                                                    "LY-CoV555",
                                                    as.character(Kds_simplified.withoutACE2$Antibody)))
  
Kds_simplified.withoutACE2$Antibody.used<-factor(Kds_simplified.withoutACE2$Antibody.used,
                                             levels = c("LY-CoV016","LY-CoV555","REGN10987","S309")
)


wuhan.Kd$Antibody.used <- ifelse(wuhan.Kd$Antibody=="CB6","LY-CoV016",
                                             ifelse(wuhan.Kd$Antibody=="CoV555",
                                                    "LY-CoV555",
                                                    as.character(wuhan.Kd$Antibody)))
  
wuhan.Kd$Antibody.used<-factor(wuhan.Kd$Antibody.used,
                                             levels = c("LY-CoV016","LY-CoV555","REGN10987","S309")
)

omicron.Kd$Antibody.used <- ifelse(omicron.Kd$Antibody=="CB6","LY-CoV016",
                                             ifelse(omicron.Kd$Antibody=="CoV555",
                                                    "LY-CoV555",
                                                    as.character(omicron.Kd$Antibody)))
  
omicron.Kd$Antibody.used<-factor(omicron.Kd$Antibody.used,
                                             levels = c("LY-CoV016","LY-CoV555","REGN10987","S309")
)

###Plot FIGURE 1A###
ggplot(Kds_simplified.withoutACE2, aes(x=Log10Kd.nb))+
  geom_histogram( position="identity",bins=40,  size = 0,alpha=1,fill="#333333")+
  labs(x=expression(-log~italic(K)[paste("D", ",", "app")])) +
  ylab("Count") +
  theme_classic()+
  #geom_vline(aes(xintercept = wuhan.Kd), colour = brewer.pal(n = 8, name = "Set1")[2],size=1.5)+
  #geom_vline(aes(xintercept = omicron.Kd), colour = brewer.pal(n = 8, name = "Set1")[1],size=1.5)+
  scale_x_continuous(breaks=c(5.15,6:11),labels=c("NB",6:11))+
  #ylim(0,16000)+
  facet_grid(.~Antibody.used)+
  geom_vline(data=wuhan.Kd[1:4,],aes(xintercept = Kd), 
             colour = brewer.pal(n = 8, name = "Set1")[2],
             linetype = "dashed",
             size=0.5)+
  geom_vline(data=omicron.Kd[1:4,],aes(xintercept = Kd), 
             colour = brewer.pal(n = 8, name = "Set1")[1],
             linetype = "dashed",
             size=0.5)+
  theme(axis.text.x = element_text(angle = 0,colour="black"),
        axis.text.y = element_text(colour="black"),
        axis.text=element_text(size=12),
        legend.text=element_text(size=12),
        legend.title=element_blank(),
        axis.title=element_text(size=14),
        panel.grid.minor.x = element_blank(),
        legend.position = "none",
        panel.spacing = unit(0.9, "lines"),
        strip.text.x = element_text(size = 12))

ggsave("figures/Figure1_histogram_distribution_of_Kds_all_abs.png", width = 7.5, height = 2, units = "in")
```

Figure 1B: Summary of Effects and Contact
```{r summary table}
Kds_effect_summarized <- Kds_mutation_effects.long.filtered %>%
  select(`KD Effect`, Mutation, Antibody)%>%
  group_by(Mutation,Antibody) %>% 
  summarize(Mean = mean(`KD Effect`,na.rm=TRUE),
            Std = sd(`KD Effect`,na.rm=TRUE))

Kds_effect_summarized$Antibody <- factor(Kds_effect_summarized$Antibody, levels=c(antibodies,"ACE2"))

Kds_effect_summarized_just_mean <- Kds_effect_summarized%>%
  select(Mutation, Antibody,Mean)%>%
  pivot_wider(names_from = Mutation, values_from = Mean)
Kds_effect_summarized_just_mean$Antibody <- factor(Kds_effect_summarized_just_mean$Antibody, levels=c(antibodies,"ACE2"))

Kds_effect_summarized$Antibody.used <- ifelse(Kds_effect_summarized$Antibody=="CB6","LY-CoV016",
                                             ifelse(Kds_effect_summarized$Antibody=="CoV555",
                                                    "LY-CoV555",
                                                    as.character(Kds_effect_summarized$Antibody)))
  
Kds_effect_summarized$Antibody.used<-factor(Kds_effect_summarized$Antibody.used,
                                             levels = c("ACE2",rev(c("LY-CoV016","LY-CoV555","REGN10987","S309")))
)
#Mean heatmap
ggplot(Kds_effect_summarized) + 
  geom_tile(mapping=aes(x=`Mutation`, y=`Antibody.used`, fill= `Mean`)) +
  scale_fill_gradient2(low = brewer.pal(8,"Set1")[2],
                       mid = "white",
                       high = brewer.pal(8,"Set1")[1],
                       midpoint = 0,
                       guide = "colorbar") + 
  labs(fill=expression(paste("Mean Effect on",-log~italic(K)[paste("D", ",", "app")]," (-log M)")))+
  xlab("Mutation") +
  ylab("Antibody or ACE2") +
  theme_classic()+
  coord_equal()+
  theme(axis.text.x = element_text(angle = 90,hjust = 1, vjust = 0.5),
        axis.text=element_text(size=12,color="black"),
        legend.text=element_text(size=10),
        legend.title=element_blank(),
        axis.title=element_text(size=14),
        legend.position="none",
        legend.key.height = unit(0.4, "in"),
        legend.justification = "top")

ggsave("figures/FigureS1_summary_table_mean_effects_edit.png", width = 6, height = 3.5, units = "in")


#Sd heatmap
ggplot(Kds_effect_summarized) + 
  geom_tile(mapping=aes(x=`Mutation`, y=`Antibody`, fill= `Std`)) +
  scale_fill_gradient2(limits = c(0.05,1.5),low = "#FFFF33",
                       high = "#121212",
                       guide = "colorbar") + 
  labs(fill="Standard Deviation of Effect")+
  xlab("Mutation") +
  ylab("Antibody") +
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90,hjust = 1, vjust = 0.5),
        axis.text=element_text(size=12,color="black"),
        #legend.text=element_text(size=12),
        legend.title=element_text(size=11,hjust = 1, vjust = 0.8),
        axis.title=element_text(size=14),
        legend.position="bottom",
        legend.spacing.x = unit(0.2, 'cm'))
ggsave("figures/FigureS1_summary_table_std_effects.png", width = 6, height = 2.5, units = "in")

#Heatmap with residues
#load structure data
first_order_structure <- read.delim("data/first_order_structure_summary.csv" ,header = TRUE, sep = ",", dec = ".")
first_order_structure$Term <- factor(rep(nn.muts,4),levels=nn.muts)
first_order_structure$Antibody <- factor(first_order_structure$Antibody,levels=antibodies)
first_order_structure$Buried_SA.nb <- ifelse(first_order_structure$Buried_SA<0, 
                                          log10(0), 
                                          log10(first_order_structure$Buried_SA))
first_order_structure$Buried_SA.nb2 <- ifelse(first_order_structure$Buried_SA<0, 
                                              0, 
                                              first_order_structure$Buried_SA)

first_order_structure$Antibody.used <- ifelse(first_order_structure$Antibody=="CB6","LY-CoV016",
                                             ifelse(first_order_structure$Antibody=="CoV555",
                                                    "LY-CoV555",
                                                    as.character(first_order_structure$Antibody)))

first_order_structure$Antibody.used <- factor(first_order_structure$Antibody.used,
                                              levels = c("S309",
                                                         "REGN10987",
                                                         "LY-CoV555",
                                                         "LY-CoV016")
)
first_order_structure <- rbind(subset(first_order_structure,Antibody=="CB6"),
                               subset(first_order_structure,Antibody=="CoV555"),
                               subset(first_order_structure,Antibody=="REGN10987"),
                               subset(first_order_structure,Antibody=="S309"))

first_order_structure$Antibody.used <- factor(first_order_structure$Antibody.used,
                                              levels=c("LY-CoV016","LY-CoV555","REGN10987","S309"))

first_order_structure$Antibody.rev <-  factor(first_order_structure$Antibody.used,
                                              levels=rev(c("LY-CoV016","LY-CoV555","REGN10987","S309")))

first_order_structure$single.effect<- Kds_effect_summarized$Mean[
  match(interaction(first_order_structure$Term,first_order_structure$Antibody),
        interaction(Kds_effect_summarized$Mutation,Kds_effect_summarized$Antibody))]

#codes to generate the triangles in the heatmap
make_triangles <- function(x, y, point = "up") {
  x <- as.integer(as.factor((x)))
  y <- as.integer(as.factor((y)))
  
  if (point == "up") {
    newx <- sapply(x, function(x) {
      c(x - 0.5, x - 0.5, x + 0.5)
    }, simplify = FALSE)
    newy <- sapply(y, function(y) {
      c(y - 0.5, y + 0.5, y + 0.5)
    }, simplify = FALSE)
  } else if (point == "down") {
    newx <- sapply(x, function(x) {
      c(x - 0.5, x + 0.5, x + 0.5)
    }, simplify = FALSE)
    newy <- sapply(y, function(y) {
      c(y - 0.5, y - 0.5, y + 0.5)
    }, simplify = FALSE)
  }
  data.frame(x = unlist(newx), y = unlist(newy))
}
newcoord_up <- make_triangles(first_order_structure$Term, first_order_structure$Antibody.rev)
newcoord_down <- make_triangles(first_order_structure$Term, first_order_structure$Antibody.rev, point = "down")
newcoord_down <- newcoord_down %>% select(xdown = x, ydown = y)
first_order_structure.new <- map_df(1:nrow(first_order_structure), function(i) first_order_structure[rep(i, 3), ])
newdata <- bind_cols(first_order_structure.new, newcoord_up, newcoord_down)

# vary mutation labels based on contact
mutation.color <- c("#876399",
                    "black",
                    "black",
                    "black",
                    "#3989BD",
                    "#C2846E",
                    "#D4A158",
                    "#3989BD",
                    "black",
                    "#729395",
                    "#729395",
                    "black",
                    "#D4A158",
                    "#3989BD",
                    "#3989BD")

newdata$Antibody.used <- factor(newdata$Antibody.used,
                                              levels=rev(c("LY-CoV016","LY-CoV555","REGN10987","S309")))


ggplot(newdata) +
  geom_polygon(aes(x = x, y = y, fill = single.effect, group = interaction(Term, Antibody.rev)), 
               color = "black",
               size=0.05) +
  scale_fill_gradient2(low = brewer.pal(8,"Set1")[2],
                       mid = "white",
                       high = brewer.pal(8,"Set1")[1],
                       midpoint = 0,
                       guide = "colorbar")+
  labs(fill=expression(paste("Mean Effect on",-log~italic(K)[paste("D", ",", "app")]," (-log M)")))+
  ggnewscale::new_scale_fill() +
  
  geom_polygon(aes(x = xdown, y = ydown, 
                   fill = Buried_SA.nb2, group = interaction(Term, Antibody.rev)), 
               color = "black",
               size=0.05) +
  scale_fill_gradient2(low = "#009E73",
                       mid = "white",
                       high = "#FFD92F",
                       midpoint = 0,
                       guide = "colorbar")+
  labs(fill=expression(paste("Contact Surface Area (",ring(A),{}^2,")")))+
  xlab("Mutation") +
  ylab("Antibody") +  
  scale_x_continuous(breaks = seq_along(unique(first_order_structure$Term)), 
                     labels = unique(levels(first_order_structure$Term))) +
  scale_y_continuous(breaks = rev(seq_along(unique(first_order_structure$Antibody.used))),
                     labels = unique(first_order_structure$Antibody.used))+
  coord_equal()+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90,hjust = 1, vjust = 0.5,color=mutation.color),
        axis.text=element_text(size=12,color="black"),
        legend.text=element_text(size=10,vjust=1),
        legend.title=element_text(size=11,vjust=0.6),
        axis.title=element_text(size=14),
        legend.position="top",
        legend.key.size = unit(0.8, "lines"),
        legend.key.width = unit(0.2, "in"),
        #legend.margin=margin(0,0,0,0),
        legend.justification = "top"
  )
ggsave("figures/Figure1_buried_surface_area_with_legend_heatmap.png", width = 8, height = 2, units = "in")
```

Figure 1D-G: Kd Distribution across Numbers of Mutations 
```{r jitter plots per ab}
###Plot FIGURE 1B###
color.antibodies <- getPalettePairedLong(15)[c(4,6,10,12)]
#CB6
ggplot(subset(Kds_simplified,Antibody=="CB6"), aes(x=Class, y = Log10Kd.nb))+
  #geom_violin() +
  geom_jitter(height = 0.15, width = 0.3, size=0.8,alpha = 0.4,aes(color = `Mutation 417`,fill=`Mutation 417`))+
  #geom_jitter(height = 0.05, width = 0.3, size=0.8,alpha = 0.4)+
  #geom_violin(width=1.4) +
  geom_boxplot(width=0.7, color="#484848",alpha=0.4, outlier.alpha = 0.0,size=0.5) +
  #scale_fill_manual(values = getYellowtoRed(16))+
  #geom_density(alpha=.2, fill="#FF6666") +
  xlab("Number of Mutations") +
  labs(y=expression(-log~italic(K)[paste("D", ",", "app")]~"LY-CoV016"),color="K417N",fill="K417N") +
  theme_classic()+
  scale_color_manual(labels=c("1" = "N", 
                            "0" = "K"),values = color.antibodies)+
  scale_fill_manual(labels=c("1" = "N", 
                            "0" = "K"),values = color.antibodies)+
  #scale_color_manual(values = color.antibodies)+
  scale_y_continuous(limits =c(5,11),breaks=c(5.2,6:11),labels=c("NB",6:11))+
  theme(axis.text.x = element_text(angle = 0,colour="black",size=11),
        axis.text.y = element_text(colour="black",size=12),
        legend.text=element_text(size=10),
        legend.title=element_text(size=10),
        legend.position = "top",
        axis.title=element_text(size=13),
        legend.spacing = unit(4, "pt"), 
        legend.key.height = unit(8, "pt"), 
        legend.spacing.x = unit(4, "pt"), 
        legend.key.width  = unit(4, "pt"), 
        legend.margin=margin(4,4,4,4),
        legend.justification = "bottom",
        legend.box.spacing = margin(0.1))+
  guides(color=guide_legend(ncol=2),fill=guide_legend(ncol=2))+
  #facet_wrap(.~Antibody)+
  geom_hline(data=subset(wuhan.Kd,Antibody=="CB6"),
             aes(yintercept = Kd),colour = "black", linetype = "dashed")

ggsave("figures/Figure1_boxplot_and_jitters_Kds_across_numb_mutations_CB6_double_fragmented.png", width = 3.75, height = 2.5, units = "in")

#REGN10987
ggplot(subset(Kds_simplified,Antibody=="REGN10987"), aes(x=Class, y = Log10Kd.nb))+
  #geom_violin() +
  geom_jitter(height = 0.15, width = 0.3, size=0.8,alpha = 0.4,aes(color = interaction(`Mutation 440`,
                                                                          `Mutation 446`,sep="/",
                                                                          lex.order=TRUE),
                                                      fill = interaction(`Mutation 440`,
                                                                          `Mutation 446`,sep="/",
                                                                          lex.order=TRUE)))+
  geom_boxplot(width=0.7, color="#484848",alpha=0.4, outlier.alpha = 0.0,size=0.5) +
  #ylim(6,11)+
  xlab("Number of Mutations") +
  labs(y=expression(-log~italic(K)[paste("D", ",", "app")]~"REGN10987"),color="N440K/G446S",fill="N440K/G446S") +
  theme_classic()+
  scale_color_manual(labels=c("0/0" = "N/G", 
                            "1/0" = "K/G",
                            "0/1" = "N/S",
                            "1/1" = "K/S"),values = color.antibodies)+
  scale_fill_manual(labels=c("0/0" = "N/G", 
                            "1/0" = "K/G",
                            "0/1" = "N/S",
                            "1/1" = "K/S"),values = color.antibodies)+
  #scale_color_manual(values = color.antibodies)+
  scale_y_continuous(limits =c(5,11),breaks=c(5.2,6:11),labels=c("NB",6:11))+
  theme(axis.text.x = element_text(angle = 0,colour="black",size=11),
        axis.text.y = element_text(colour="black",size=12),
        legend.text=element_text(size=10),
        legend.title=element_text(size=10),
        legend.position = "top",
        axis.title=element_text(size=13),
        legend.spacing = unit(4, "pt"), 
        legend.key.height = unit(8, "pt"), 
        legend.spacing.x = unit(4, "pt"), 
        legend.key.width  = unit(4, "pt"), 
        legend.margin=margin(4,4,4,4),
        legend.justification = "bottom",
        legend.box.spacing = margin(0.1))+
  guides(color=guide_legend(ncol=2),fill=guide_legend(ncol=2))+
  #facet_wrap(.~Antibody)+
  geom_hline(data=subset(wuhan.Kd,Antibody=="REGN10987"),
             aes(yintercept = Kd),colour = "black", linetype = "dashed")

ggsave("figures/Figure1_boxplot_and_jitters_Kds_across_numb_mutations_REGN10987_double_fragmented.png", width = 3.75, height = 2.5, units = "in")

#CoV555
ggplot(subset(Kds_simplified,Antibody=="CoV555"), aes(x=Class, y = Log10Kd.nb))+
  #geom_violin() +
  geom_jitter(height = 0.15, width = 0.3, size=0.8,alpha = 0.4,aes(color = interaction(`Mutation 484`,
                                                                          `Mutation 493`,sep="/",
                                                                          lex.order=TRUE),
                                                      fill = interaction(`Mutation 484`,
                                                                          `Mutation 493`,sep="/",
                                                                          lex.order=TRUE)))+
  geom_boxplot(width=0.7, color="#484848",alpha=0.4, outlier.alpha = 0.0,size=0.5) +
  #ylim(6,11)+
  xlab("Number of Mutations") +
  labs(y=expression(-log~italic(K)[paste("D", ",", "app")]~"LY-CoV555"),color="E484A/Q493R",fill="E484A/Q493R") +
  theme_classic()+
  scale_color_manual(labels=c("0/0" = "E/Q", 
                            "1/0" = "A/Q",
                            "0/1" = "E/R",
                            "1/1" = "A/R"),values = color.antibodies)+
  scale_fill_manual(labels=c("0/0" = "E/Q", 
                            "1/0" = "A/Q",
                            "0/1" = "E/R",
                            "1/1" = "A/R"),values = color.antibodies)+
  #scale_color_manual(values = color.antibodies)+
  scale_y_continuous(limits =c(5,11),breaks=c(5.2,6:11),labels=c("NB",6:11))+
  theme(axis.text.x = element_text(angle = 0,colour="black",size=11),
        axis.text.y = element_text(colour="black",size=12),
        legend.text=element_text(size=10),
        legend.title=element_text(size=10),
        legend.position = "top",
        axis.title=element_text(size=13),
        legend.spacing = unit(4, "pt"), 
        legend.key.height = unit(8, "pt"), 
        legend.spacing.x = unit(4, "pt"), 
        legend.key.width  = unit(4, "pt"), 
        legend.margin=margin(4,4,4,4),
        legend.justification = "bottom",
        legend.box.spacing = margin(0.1))+
  guides(color=guide_legend(ncol=2),fill=guide_legend(ncol=2))+
  #facet_wrap(.~Antibody)+
  geom_hline(data=subset(wuhan.Kd,Antibody=="CoV555"),
             aes(yintercept = Kd),colour = "black", linetype = "dashed")

ggsave("figures/Figure1_boxplot_and_jitters_Kds_across_numb_mutations_CoV555_double_fragmented.png", width = 3.75, height = 2.5, units = "in")

#S309
ggplot(subset(Kds_simplified,Antibody=="S309"), aes(x=Class, y = Log10Kd.nb))+
  geom_jitter(height = 0.05, width = 0.3, size=0.8,alpha = 0.4,color=color.antibodies[1])+
  #geom_jitter(height = 0.05, width = 0.3, size=0.8,alpha = 0.4,color=color.antibodies[1],fill=color.antibodies[1])+
  geom_boxplot(width=0.7, color="#484848",alpha=0.4, outlier.alpha = 0.0,size=0.5) +
  xlab("Number of Mutations") +
  labs(y=expression(-log~italic(K)[paste("D", ",", "app")]~"S309")) +
  theme_classic()+
  #scale_color_manual(values = color.antibodies)+
  scale_y_continuous(limits =c(5,11),breaks=c(5.2,6:11),labels=c("NB",6:11))+
  theme(axis.text.x = element_text(angle = 0,colour="black",size=11),
        axis.text.y = element_text(colour="black",size=12),
        legend.text=element_text(size=10),
        legend.title=element_text(size=10),
        legend.position = "top",
        axis.title=element_text(size=13),
        legend.spacing = unit(4, "pt"), 
        legend.key.height = unit(8, "pt"), 
        legend.spacing.x = unit(4, "pt"), 
        legend.key.width  = unit(4, "pt"), 
        legend.margin=margin(4,4,4,4),
        legend.justification = "bottom",
        legend.box.spacing = margin(0.1))+
  #facet_wrap(.~Antibody)+
  #ylim(6,11)+
  #scale_y_continuous(breaks=6:11,labels=c("NB",7:11))+
  geom_hline(data=subset(wuhan.Kd,Antibody=="S309"),
             aes(yintercept = Kd),colour = "black", linetype = "dashed")

ggsave("figures/Figure1_boxplot_and_jitters_Kds_across_numb_mutations_S309_double_fragmented.png", width = 3.75, height = 2.5, units = "in")
```

##Figure 2 - Escape Mutations
Set Up and Figure 2B: Non-Binding Effects
```{r sufficient mutations}
Kds_mutation_effects.long.filtered <- subset(Kds_mutational_effects.long.all.abs,
                                             `KD WT` > 6)
binding.percentage <- data.frame(Mutation = nn.muts)

get.sum.binding <- function(x,ab){
  with(subset(Kds_mutation_effects.long.filtered,Antibody==ab),sum(Mutation==x & `KD OM` == 6)/sum(Mutation==x))
}

get.sum.deleterious <- function(x,ab){
  with(subset(Kds_mutation_effects.long.filtered,Antibody==ab),sum(Mutation==x & `KD Effect` <0)/sum(Mutation==x))
  #with(subset(Kds_mutation_effects.long.filtered,Antibody==ab),
      # sum(Mutation==x & `KD Effect` <0 & (abs(`KD Effect`) + 0.05)> `Effect Error`,na.rm=TRUE)/sum(Mutation==x)) #only significant
}

binding.percentage$S309 <- apply(binding.percentage,1,get.sum.binding,ab="S309")
binding.percentage$CoV555 <- apply(binding.percentage,1,get.sum.binding,ab="CoV555")
binding.percentage$CB6 <- apply(binding.percentage,1,get.sum.binding,ab="CB6")
binding.percentage$REGN10987 <- apply(binding.percentage,1,get.sum.binding,ab="REGN10987")

deleterious.percentage <- data.frame(Mutation = nn.muts)
deleterious.percentage$S309 <- apply(deleterious.percentage,1,get.sum.deleterious,ab="S309")
deleterious.percentage$CoV555 <- apply(deleterious.percentage,1,get.sum.deleterious,ab="CoV555")
deleterious.percentage$CB6 <- apply(deleterious.percentage,1,get.sum.deleterious,ab="CB6")
deleterious.percentage$REGN10987 <- apply(deleterious.percentage,1,get.sum.deleterious,ab="REGN10987")

binding.percentage.long <- reshape(binding.percentage,direction="long",varying=list(antibodies))
binding.percentage.long$Antibody <- rep(antibodies,each=15)
binding.percentage.long$Mutation <- factor(binding.percentage.long$Mutation,levels=nn.muts)

deleterious.percentage.long <- reshape(deleterious.percentage,direction="long",varying=list(antibodies))
deleterious.percentage.long$Antibody <- rep(antibodies,each=15)
deleterious.percentage.long$Mutation <- factor(deleterious.percentage.long$Mutation,levels=nn.muts)

ggplot(deleterious.percentage.long, aes(x=Mutation, y=CB6*100, fill = Antibody))+
  geom_bar(stat='identity',color="#333333",position = position_dodge())+
  #geom_errorbar(aes(ymin=Coefficient-Standard.Error, ymax=Coefficient+Standard.Error), width=.6,
  #              position=position_dodge(.9), colour = "#969696",size=0.7) +
  scale_color_manual(values=getPalettePairedLong(6)[2:5])+
  geom_hline(yintercept=50,size=0.5,linetype = "dashed",color="black")+
  xlab("Mutation") +
  ylab(str_wrap("Percent Deleterious Effects", width = 30)) +
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90,hjust = 1, vjust = 0.5),
        axis.text=element_text(size=12,color="black"),
        legend.text=element_text(size=12),
        legend.title=element_blank(),
        axis.title=element_text(size=14),
        axis.title.y=element_text(size=12))

ggsave("figures/Figure2SI_bars_percent_deleterious_effects_binders_only_significant.png", width = 7.5, height = 3.5, units = "in")

ggplot(subset(binding.percentage.long,Antibody!= "S309"), aes(x=Mutation, y=CB6*100, fill = Antibody))+
  geom_bar(stat='identity',color="#333333",width=0.7,position=position_dodge(0.7))+
  #geom_errorbar(aes(ymin=Coefficient-Standard.Error, ymax=Coefficient+Standard.Error), width=.6,
  #              position=position_dodge(.9), colour = "#969696",size=0.7) +
  scale_fill_manual(values=getPalettePairedLong(6)[2:5])+
  #geom_hline(yintercept=0.05,size=2,color="white")+
  xlab("Mutation") +
  ylab(str_wrap("Percent Non-Binding Effects", width = 25)) +
  theme_classic()+
  guides(fill = guide_legend(override.aes = list(size = 0.2)),
         color = guide_legend(override.aes = list(size = 0.2)))+
  theme(axis.text.x = element_text(angle = 90,hjust = 1, vjust = 0.5),
        axis.text=element_text(size=12,color="black"),
        legend.text=element_text(size=8),
        legend.title=element_blank(),
        legend.position = "none",
        axis.title=element_text(size=14),
        axis.title.y=element_text(size=12),
        legend.key.size = unit(0.8, "lines"))

ggsave("figures/Figure2_bars_percent_non_effect_binders_is_it_sufficient.png", width = 3.75, height = 2.5, units = "in")
```

Figure 2A: Frequency amongst Non-Binders
```{r required mutations}
Kds.nonbinders <- subset(Kds_simplified,Log10Kd==6)
get.mean.nonbinders <- function(df){
  x <- df["Mutation"]
  ab <- df["Antibody"]
  mat <- subset(Kds.nonbinders,Antibody==ab)
  row <- mat[paste0("Mutation ",x)]
  mean(row=='1')
}

nonbinding.percentage <- data.frame(Mutation = factor(rep(as.character(nn),4),levels=nn),
                                    Antibody = factor(rep(antibodies,each=15),levels=antibodies),
                                    Antibody.used = factor(rep(c("LY-CoV016","LY-CoV555","REGN10987","S309"),
                                                          each=15),
                                                      levels=c("LY-CoV016","LY-CoV555","REGN10987","S309")))


nonbinding.percentage$Proportion <- apply(nonbinding.percentage,1,get.mean.nonbinders)
nonbinding.percentage$Mutation <- factor(rep(nn.muts,4),levels = nn.muts)

ggplot(subset(nonbinding.percentage,Antibody.used != "S309"), aes(x=Mutation, y=Proportion*100, fill = Antibody.used))+
  geom_bar(stat='identity',color="#333333",width=0.7,position=position_dodge(0.7))+
  #geom_errorbar(aes(ymin=Coefficient-Standard.Error, ymax=Coefficient+Standard.Error), width=.6,
  #              position=position_dodge(.9), colour = "#969696",size=0.7) +
  scale_fill_manual(values=getPalettePairedLong(6)[2:5])+
  geom_hline(yintercept=50,size=0.5,linetype = "dashed",color="black")+
  xlab("Mutation") +
  ylab(str_wrap("Frequency amongst Non-Binders", width = 20)) +
  theme_classic()+
  guides(fill = guide_legend(override.aes = list(size = 0.15),ncol=2),
         color = guide_legend(override.aes = list(size = 0.15),ncol=2))+
  theme(axis.text.x = element_text(angle = 90,hjust = 1, vjust = 0.5),
        axis.text=element_text(size=12,color="black"),
        legend.text=element_text(size=8),
        legend.title=element_blank(),
        legend.position = c(.77,0.9),
        axis.title=element_text(size=14),
        axis.title.y=element_text(size=12),
        legend.key.size = unit(0.4, "lines"))

ggsave("figures/Figure2_bars_percent_mutated_in_non_binders_required.png", width = 4, height = 2.5, units = "in")

get.nonbinder.percentage.per.mutation <- function(df){
  x <- df["Mutation"]
  ab <- df["Antibody"]
  mat <- subset(Kds_simplified,Antibody==ab)
  row <- subset(mat,get(paste0("Mutation ",x))=='1')['Log10Kd']
  mean(row==6)
}

nonbinder.per.mutations <- data.frame(Mutation = factor(rep(as.character(nn),4),levels=nn),
                                    Antibody = factor(rep(antibodies,each=15),levels=antibodies))


nonbinder.per.mutations$Proportion <- apply(nonbinder.per.mutations,1,get.nonbinder.percentage.per.mutation)
nonbinder.per.mutations <- subset(nonbinder.per.mutations,Antibody != "S309")

ggplot(nonbinder.per.mutations, aes(x=Mutation, y=Proportion*100, fill = Antibody))+
  geom_bar(stat='identity',color="#333333",position = position_dodge())+
  #geom_errorbar(aes(ymin=Coefficient-Standard.Error, ymax=Coefficient+Standard.Error), width=.6,
  #              position=position_dodge(.9), colour = "#969696",size=0.7) +
  scale_fill_manual(values = antibody.colors)+
  geom_hline(yintercept=50,size=0.7,linetype = "dashed",color="black")+
  geom_hline(yintercept=25,size=0.7,linetype = "dotted",color="black")+

  xlab("Mutation") +
  ylab(str_wrap("Percent Non-Binders when Mutated", width = 30)) +
  theme_classic()+
  guides(fill = guide_legend(override.aes = list(size = 0.5)))+
  theme(axis.text.x = element_text(angle = 90,hjust = 1, vjust = 0.5),
        axis.text=element_text(size=12,color="black"),
        legend.text=element_text(size=12),
        legend.title=element_blank(),
        axis.title=element_text(size=14),
        axis.title.y=element_text(size=12))

ggsave("figures/FigureSI2_bars_percent_non_binders_when_mutated.png", width = 7.5, height = 3.5, units = "in")
```

Figure 2C-...: Decision tree
```{r decision tree}
library(rpart)       # direct engine for decision tree application
library(caret)       # meta engine for decision tree application

# Model interpretability packages
library(rpart.plot)  # for plotting decision trees
library(vip)         # for feature importance
library(pdp)         # for feature effects
library(treeheatr)
library(partykit)
library(rattle)
library(ggparty)

#Kds_simplified[paste0('Mutation_', nn)] <- Kds_simplified[paste0('Mutation ', nn)]
#Kds_simplified[nn.muts] <- Kds_simplified[paste0('Mutation ', nn)]
Kds.abs <- subset(Kds_simplified,Antibody=="CoV555")
Kds.abs.CoV555 <- subset(Kds_simplified,Antibody=="CoV555" & (NonBinder | Filtered))
Kds.abs.S309 <- subset(Kds_simplified,Antibody=="S309" & (NonBinder |Filtered))
Kds.abs.CB6 <- subset(Kds_simplified,Antibody=="CB6" & (NonBinder |Filtered))
Kds.abs.REGN10987 <- subset(Kds_simplified,Antibody=="REGN10987" & (NonBinder |Filtered))

Kds.muts.REGN10987 <- select(Kds.abs.REGN10987,c('Log10Kd',nn.muts))
Kds.muts.CB6 <- select(Kds.abs.CB6,c('Log10Kd',nn.muts))
Kds.muts.CoV555 <- select(Kds.abs.CoV555,c('Log10Kd',nn.muts))
Kds.muts.S309 <- select(Kds.abs.S309,c('Log10Kd',nn.muts))

Kds.bind.REGN10987 <- Kds.muts.REGN10987
Kds.bind.CB6 <- Kds.muts.CB6
Kds.bind.CoV555 <- Kds.muts.CoV555
Kds.bind.S309 <- Kds.muts.S309

Kds.bind.REGN10987$Binds <- as.factor(ifelse(Kds.bind.REGN10987$Log10Kd == 6, "Non-Binding", "Binding"))
Kds.bind.CB6$Binds <- as.factor(ifelse(Kds.bind.CB6$Log10Kd == 6, "Non-Binding", "Binding"))
Kds.bind.CoV555$Binds <- as.factor(ifelse(Kds.bind.CoV555$Log10Kd == 6, "Non-Binding", "Binding"))
Kds.bind.S309$Binds <- as.factor(ifelse(Kds.bind.S309$Log10Kd == 6, "Non-Binding", "Binding"))

#decision.tree.S309 <- rpart(
#  formula = Binds ~ G339D + S371L + S373P + S375F + K417N + N440K + G446S + 
#    S477N + T478K + E484A + Q493R + G496S + Q498R + N501Y + Y505H + 0*Log10Kd,
#  data    = Kds.bind.S309
#)

decision.tree.CoV555 <- rpart(
  formula = Binds ~ G339D + S371L + S373P + S375F + K417N + N440K + G446S + 
    S477N + T478K + E484A + Q493R + G496S + Q498R + N501Y + Y505H + 0*Log10Kd,
  data    = Kds.bind.CoV555
  #method  = "anova"
)
decision.tree.CB6 <- rpart(
  formula = Binds ~ G339D + S371L + S373P + S375F + K417N + N440K + G446S + 
    S477N + T478K + E484A + Q493R + G496S + Q498R + N501Y + Y505H + 0*Log10Kd,
  data    = Kds.bind.CB6
)
decision.tree.REGN10987 <- rpart(
  formula = Binds ~ G339D + S371L + S373P + S375F + K417N + N440K + G446S + 
    S477N + T478K + E484A + Q493R + G496S + Q498R + N501Y + Y505H + 0*Log10Kd,
  data    = Kds.bind.REGN10987
)

decision.tree.CB6.party <- as.party(decision.tree.CB6)
decision.tree.REGN10987.party <- as.party(decision.tree.REGN10987)
decision.tree.CoV555.party <- as.party(decision.tree.CoV555)
#decision.tree.S309.party <- as.party(decision.tree.S309) 
tiff("figures/CoV55_decision_tree.tiff", units="in", width = 8, height = 4.8, res = 300)
ggparty(decision.tree.CoV555.party,terminal_space = 0.4) +
  geom_edge() +
  geom_edge_label(size = 3) +
  geom_node_label(aes(label = splitvar),
                  ids = "inner",
                  size=3) +
  geom_node_plot(gglist = list(geom_bar(aes(x = "", fill = Binds),
                                        position = position_fill()),
                               coord_flip(),
                               theme_void(),
                               scale_fill_manual(values=c("#3879A6","#C63637")),
                               theme(legend.position = "none")),
                 #siz = "nodesize",
                 shared_legend = FALSE,
                 height = 0.05)+
  theme(legend.position = "none")+
  geom_node_plot(gglist = list(geom_histogram(aes(x = Log10Kd)),
                               coord_flip(),
                               theme_bw(),
                               ylab("Count"),
                               scale_x_continuous(limits=c(5.9,10.5),breaks=c(6:10),labels=c("NB","7","8","9","10")),
                               scale_y_log10(breaks=c(1,10,100,1000,10000),labels=c(1,"",100,"",10000),position="left"),
                               labs(x=expression(-log~italic(K)[paste("D", ",", "app")])),
                               theme(axis.text.x = element_text(angle = 270,colour="black",hjust = 0, vjust = 0.5,size=8),
                                     axis.text.y = element_text(colour="black",size=10),
                                     #axis.text=element_text(size=10),
                                     legend.text=element_text(size=12),
                                     legend.title=element_blank(),
                                     axis.title=element_text(size=12),
                                     axis.title.y = element_text(angle=90),
                                     panel.grid.minor.x = element_blank(),
                                     legend.position = "none",
                                     panel.spacing = unit(0.7, "lines"),
                                     strip.text.x = element_blank())),
                 # draw only one label for each axis
                 shared_axis_labels = TRUE,
                 height = 0.95,
                 nudge_y= -0.025
                 # draw line between tree and legend
                 #legend_separator = TRUE
  )
dev.off()

tiff("figures/REGN10987_decision_tree.tiff", units="in", width = 6, height = 4, res = 300)
ggparty(decision.tree.REGN10987.party,terminal_space = 0.4) +
  geom_edge() +
  geom_edge_label(size = 3) +
  geom_node_label(aes(label = splitvar),
                  ids = "inner",
                  size=3) +
  geom_node_plot(gglist = list(geom_bar(aes(x = "", fill = Binds),
                                        position = position_fill()),
                               coord_flip(),
                               theme_void(),
                               scale_fill_manual(values=c("#3879A6","#C63637")),
                               theme(legend.position = "none")),
                 #siz = "nodesize",
                 shared_legend = FALSE,
                 height = 0.05)+
  theme(legend.position = "none")+
  geom_node_plot(gglist = list(geom_histogram(aes(x = Log10Kd)),
                               coord_flip(),
                               theme_bw(),
                               ylab("Count"),
                               scale_x_continuous(limits=c(5.9,10.5),breaks=c(6:10),labels=c("NB","7","8","9","10")),
                               scale_y_log10(breaks=c(1,10,100,1000,10000),labels=c(1,"",100,"",10000),position="left"),
                               labs(x=expression(-log~italic(K)[paste("D", ",", "app")])),
                               theme(axis.text.x = element_text(angle = 270,colour="black",hjust = 0, vjust = 0.5,size=8),
                                     axis.text.y = element_text(colour="black",size=10),
                                     #axis.text=element_text(size=10),
                                     legend.text=element_text(size=12),
                                     legend.title=element_blank(),
                                     axis.title=element_text(size=12),
                                     axis.title.y = element_text(angle=90),
                                     panel.grid.minor.x = element_blank(),
                                     legend.position = "none",
                                     panel.spacing = unit(0.7, "lines"),
                                     strip.text.x = element_blank())),
                 # draw only one label for each axis
                 shared_axis_labels = TRUE,
                 height = 0.95,
                 nudge_y= -0.025
                 # draw line between tree and legend
                 #legend_separator = TRUE
  )
dev.off()

tiff("figures/CB6_decision_tree.tiff", units="in", width = 2, height = 4, res = 300)
ggparty(decision.tree.CB6.party,terminal_space = 0.4) +
  geom_edge() +
  geom_edge_label(size = 3) +
  geom_node_label(aes(label = splitvar),
                  ids = "inner",
                  size=3) +
  geom_node_plot(gglist = list(geom_bar(aes(x = "", fill = Binds),
                                        position = position_fill()),
                               coord_flip(),
                               theme_void(),
                               scale_fill_manual(values=c("#3879A6","#C63637")),
                               theme(legend.position = "none")),
                 #siz = "nodesize",
                 shared_legend = FALSE,
                 height = 0.05)+
  theme(legend.position = "none")+
  geom_node_plot(gglist = list(geom_histogram(aes(x = Log10Kd)),
                               coord_flip(),
                               theme_bw(),
                               ylab("Count"),
                               scale_x_continuous(limits=c(5.9,10.5),breaks=c(6:10),labels=c("NB","7","8","9","10")),
                               scale_y_log10(breaks=c(1,10,100,1000,10000),labels=c(1,"",100,"",10000),position="left"),
                               labs(x=expression(-log~italic(K)[paste("D", ",", "app")])),
                               theme(axis.text.x = element_text(angle = 270,colour="black",hjust = 0, vjust = 0.5,size=8),
                                     axis.text.y = element_text(colour="black",size=10),
                                     #axis.text=element_text(size=10),
                                     legend.text=element_text(size=12),
                                     legend.title=element_blank(),
                                     axis.title=element_text(size=12),
                                     axis.title.y = element_text(angle=90,vjust=-0.7),
                                     panel.grid.minor.x = element_blank(),
                                     legend.position = "none",
                                     panel.spacing = unit(0.7, "lines"),
                                     strip.text.x = element_blank())),
                 # draw only one label for each axis
                 shared_axis_labels = TRUE,
                 height = 0.95,
                 nudge_y= -0.025
                 # draw line between tree and legend
                 #legend_separator = TRUE
  )
dev.off()
```

## Figure 3 - Epistasis and Alternative Pathways
Figure 3A: Choosing the best model
```{r best model}
models.df <- read.delim("data/CV_rsquared.csv" ,skip = 0,header = TRUE, sep = ",", dec = ".",
                        colClasses=c("factor","numeric","numeric","numeric"))

models.df$Type <- factor(models.df$Type,levels=c("LY-CoV016","LY-CoV555","REGN10987","S309","ACE2"))
ggplot(models.df, aes(x=Order, y=Mean, color = Type))+
  geom_point(size=3,alpha=0.7)+
  #scale_color_manual(values=c(antibody.colors,"#121212"))+
  geom_errorbar(aes(ymin=Mean-Std, ymax=Mean+Std), width=.2,colour = "#666666") +
  xlab("Order of Epistatic Model") +
  ylab("Correlation Coefficient") +
  geom_smooth(method = "loess", se = FALSE, linetype = "dashed")+
  scale_x_continuous(breaks=1:6)+
  scale_color_manual(values=c(antibody.colors,"#121212"))+
  theme_classic()+
  #geom_hline(aes(yintercept = 9.908776e-01), colour = "gray")+
  theme(axis.text.x = element_text(angle = 0),
        axis.text=element_text(size=12),
        legend.text=element_text(size=11),
        legend.title=element_blank(),
        axis.title=element_text(size=14),
        legend.position = c(.6,.17))+
  guides(colour = guide_legend(nrow = 3))

ggsave("figures/Figure3_choosing_model.png", width = 3, height = 4, units = "in")

```

Figure SI: epistasis analyses
```{r epistasis coefficients}
library(ggtext)
library(plyr)
library(ggplot2)
library(RColorBrewer)
library(stringr)
library(gtools)
library(ggpmisc)
library(tidyr)
library(dplyr)
library(ggnewscale)
library(ggrepel)
library(paletteer)
library(cowplot)
library(Rmisc)
library(scales)

CoV555 <- read.delim("data/lasso_version_CoV555_aic_3_order_biochem.txt" ,skip = 1,header = TRUE, sep = "\t", dec = ".")#[-1,]
S309<- read.delim("data/thomas_version_S309_aic_4_order_biochem.txt" ,skip = 1,header = TRUE, sep = "\t", dec = ".")#[-1,]
CB6 <- read.delim("data/lasso_version_CB6_aic_4_order_biochem.txt" ,skip = 1,header = TRUE, sep = "\t", dec = ".")#[-1,]
REGN10987 <- read.delim("data/lasso_version_REGN10987_aic_2_order_biochem.txt" ,skip = 1,header = TRUE, sep = "\t", dec = ".")#[-1,]
ACE2 <- read.delim("data/ACE2_5_order_biochem.txt" ,skip = 1,header = TRUE, sep = "\t", dec = ".")#[-1,]

epistasis.biochem <- AppendMe(c("CB6","CoV555","REGN10987","S309","ACE2"))
#epistasis.stat <- read.delim("data/ACE2_4order_stat.txt",skip = 2,header = TRUE, sep = "\t", dec = ".")
epistasis.biochem$Order <- ifelse(epistasis.biochem$Term == "Intercept",0,str_count(epistasis.biochem$Term, pattern = ",")+1)
#epistasis.stat$Order <- ifelse(epistasis.stat$Term == "Intercept",0,str_count(epistasis.stat$Term, pattern = ",")+1)
epistasis.biochem <- separate(data = epistasis.biochem, col = Term, into = c("Term 1", "Term 2", "Term 3", "Term 4", "Term 5"), sep = ",")
epistasis.biochem[,1:5] <- apply(epistasis.biochem[,1:5],2,mapvalues,from=n,to=nn.muts)
epistasis.biochem[,1:5]<- lapply(epistasis.biochem[,1:5], function(X){ordered(X,levels = nn.muts,labels=nn.muts)})

antibody.colors  <- getPalettePairedLong(6)[2:5]

#bunch of rearranging the terms and coefficients
summing_coefficient <- function(x,y){
  return(with(y, sum(Coefficient.nb[
    rowSums(x[1] == y,na.rm=TRUE) > 0  &
      rowSums(x[2] == y,na.rm=TRUE) > 0])
  ))
}

summing_coefficient <- function(x,y){
  return(with(y, sum(Coefficient.nb[
    rowSums(x[1] == y,na.rm=TRUE) > 0  &
      rowSums(x[2] == y,na.rm=TRUE) > 0])
  ))
}
abs_summing_coefficient <- function(x,y){
  return(with(y, sum(abs(Coefficient.nb[
    rowSums(x[1] == y,na.rm=TRUE) > 0  &
      rowSums(x[2] == y,na.rm=TRUE) > 0]))
  ))
}

summing_coefficient_to_one <- function(x,y){
  return(with(y, sum(Coefficient.nb[
    rowSums(x[1] == y,na.rm=TRUE) > 0 ],na.rm=TRUE)
  ))
}
epistasis.biochem$p.value <- abs(epistasis.biochem$Standard.Error)/abs(epistasis.biochem$Coefficient)
epistasis.biochem$Coefficient.nb <- ifelse(epistasis.biochem$p.value > 1 & epistasis.biochem$Order > 1,
                                           0, epistasis.biochem$Coefficient)  #ifelse(epistasis.biochem$p.value > 0.5 & epistasis.biochem$Order > 1,0, epistasis.biochem$Coefficient)
epistasis.biochem.higher <- subset(epistasis.biochem,Order %in% c("3","4")) #higher order than 2
#epistasis.biochem.higher$one <- 1
epistasis.biochem.temp <- subset(epistasis.biochem,Order%in%c("1","2")) #only 2nd order

for(i in antibodies){
  epistasis.temp.higher<-subset(epistasis.biochem.higher,Antibody==i)
  epistasis.temp <- subset(epistasis.biochem.temp,Antibody==i)
  epistasis.temp.temp <- subset(epistasis.temp,Order=="2")
  
  epistasis.temp.temp$t1 <- epistasis.temp.temp$`Term 2`
  epistasis.temp.temp$t2 <- epistasis.temp.temp$`Term 1`
  epistasis.temp.temp$part <- "A"
  epistasis.temp.temp$second <- "Pairwise"
  
  
  epistasis.temp.temp2 <- subset(epistasis.temp.temp,Order=="2")
  epistasis.temp.temp2$`Term 1` <- epistasis.temp.temp2$t1
  epistasis.temp.temp2$`Term 2` <- epistasis.temp.temp2$t2
  #epistasis.temp.temp2$`Coefficient`<- apply(epistasis.temp.temp2,1,summing_coefficient,y=epistasis.temp.higher)
  epistasis.temp.temp2$part <- "B"
  epistasis.temp.temp2$second <- "Pairwise"
  
  epistasis.temp.temp3 <- subset(epistasis.temp,Order=="1")
  epistasis.temp.temp3$`Term 2` <- epistasis.temp.temp3$`Term 1`
  epistasis.temp.temp3$t1 <- NA
  epistasis.temp.temp3$t2 <- NA
  epistasis.temp.temp3$Coefficient.nb <- NA
  epistasis.temp.temp3$part <- "C"
  epistasis.temp.temp3$second <- "Pairwise"
  
  epistasis.temp.temp4 <- subset(epistasis.temp,Order=="1")
  epistasis.temp.temp4$`Term 2` <- "Higher"
  epistasis.temp.temp4$t1 <- NA
  epistasis.temp.temp4$t2 <- NA
  epistasis.temp.temp4$Coefficient.nb <-
    apply(epistasis.temp.temp4,1,summing_coefficient_to_one,y=epistasis.temp.higher)
  epistasis.temp.temp4$part <- "D"
  epistasis.temp.temp4$second <- ">2"
  
  epistasis.temp.df2 <- rbind(epistasis.temp.temp,epistasis.temp.temp2,epistasis.temp.temp3,epistasis.temp.temp4)
  epistasis.temp.df2$second <- factor(epistasis.temp.df2$second,levels=c("Pairwise",">2"))
  epistasis.temp.df2$`Term 2` <- factor(epistasis.temp.df2$`Term 2`,levels=c(nn.muts,"Higher"))
  epistasis.temp.df2$Antibody <- i
  assign(paste0("epistasis.biochem.df.all.",i),epistasis.temp.df2)
}
epistasis.biochem.df.all <- rbind(epistasis.biochem.df.all.CoV555,epistasis.biochem.df.all.S309,
                                  epistasis.biochem.df.all.CB6,epistasis.biochem.df.all.REGN10987)


###PLOTS NOWW
biochem.first.order <- subset(epistasis.biochem,Order=="1")
biochem.first.order$Coefficient.nb2 <- ifelse(biochem.first.order$Coefficient.nb<(-0.5),
                                              (biochem.first.order$Coefficient.nb+0.5)*0.2-.5,
                                              biochem.first.order$Coefficient.nb)
biochem.first.order$Error.nb2 <- ifelse(biochem.first.order$Coefficient.nb<(-.5),
                                        biochem.first.order$Standard.Error*0.2,
                                        biochem.first.order$Standard.Error)

ggplot(biochem.first.order, aes(x=`Term 1`, y=Coefficient.nb2, fill = Antibody))+
  geom_bar(stat='identity',color="#333333",position = position_dodge())+
  geom_errorbar(aes(ymin=Coefficient.nb2-Error.nb2, ymax=Coefficient.nb2+Error.nb2), width=.6,
                position=position_dodge(.9), colour = "#969696",size=0.7) +
  scale_fill_manual(values=getPalettePairedLong(6)[2:5])+
  #scale_fill_manual(values = antibody.colors)+
  #scale_y_continuous(trans = squash_axis(5,95,10))+
  scale_y_continuous(breaks = c(-1.8,-1.4,-1,-0.6,-0.3,0,0.3),
                     labels = c("-7","-5","-3","-1","-0.3","0","0.3"),
                     #                   trans = squash_axis(-0.8,-0.4,10)
  )+
  #geom_segment(x=-2,xend=15,y=-0.3,yend=-0.3,size=0.5,color="white")+
  geom_hline(yintercept=-0.5,size=2,color="white")+
  xlab("Mutation") +
  ylab("Linear Effect") +
  theme_classic()+
  guides(fill = guide_legend(override.aes = list(size = 0.1)),
         color = guide_legend(override.aes = list(size = 0.1)))+
  theme(axis.text.x = element_text(angle = 90,hjust = 1, vjust = 0.5),
        axis.text=element_text(size=12,color="black"),
        legend.text=element_text(size=10),
        legend.position = c(0.87,0.22),
        legend.title=element_blank(),
        axis.title=element_text(size=14),
        legend.key.size = unit(0.5, "lines"))
ggsave("figures/Figure2_bars_linear_effects_combined_with break.png", width = 4.5, height = 3.5, units = "in")
ggsave("figures/THOMAS_TOBIT_Figure2_bars_linear_effects_combined_with break.png", width = 4.5, height = 3.5, units = "in")

biochem.first.order$Antibody.used <- ifelse(biochem.first.order$Antibody=="CB6","LY-CoV016",
                                            ifelse(biochem.first.order$Antibody=="CoV555",
                                                   "LY-CoV555",
                                                   as.character(biochem.first.order$Antibody)))

biochem.first.order$Antibody.used<-factor(biochem.first.order$Antibody.used,
                                          levels = rev(c("LY-CoV016","LY-CoV555","REGN10987","S309","ACE2"))
)



ggplot(biochem.first.order) + 
  geom_tile(mapping=aes(x=`Term 1`, y=`Antibody.used`, fill= `Coefficient.nb2`)) +
  scale_fill_gradient2(breaks = c(-1.8,-1.4,-1,-0.6,-0.3,0,0.3),
                       labels = c("-7","-5","-3","-1","-0.3","0","0.3"),
                       low = brewer.pal(8,"Set1")[2],
                       mid = "white",
                       high = brewer.pal(8,"Set1")[1],
                       midpoint = 0,
                       guide = "colorbar") + 
  labs(fill="First-Order\nCoefficients")+
  xlab("Mutation") +
  ylab("Antibody") +
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90,hjust = 1, vjust = 0.5),
        axis.text=element_text(size=12,color="black"),
        legend.text=element_text(size=10,angle=0,hjust=0,vjust=0.5),
        legend.title=element_text(size=11,vjust=0.75),
        axis.title=element_text(size=14),
        legend.position="right",
        legend.key.height = unit(0.28, "in"),
        legend.justification = "top")
ggsave("figures/Figure3_summary_table_linear.png", width = 7.5, height = 2.4, units = "in")

#S309
#epistasis.biochem.df.all$Coefficient.nb2 <- ifelse(epistasis.biochem.df.all$Coefficient.nb < -0.75,
#                                                   -0.75,
#                                                   epistasis.biochem.df.all$Coefficient.nb)
for(i in antibodies){
  epistasis.biochem.df2 <- subset(epistasis.biochem.df.all,Antibody==i)
  lim.higher.low <- ifelse(i=="CoV555",-6.5,ifelse(i=="REGN10987",-1,-1))
  lim.higher.high <- ifelse(i=="CoV555",1,ifelse(i=="REGN10987",1,1))
  
  lim.2.low <- ifelse(i=="CoV555",-1,ifelse(i=="REGN10987",-4,-1))
  lim.2.high <- ifelse(i=="CoV555",1.5,ifelse(i=="CB6",1,1))
  
  ggplot() + 
    geom_tile(data = subset(epistasis.biochem.df2,second =="Pairwise"),
              mapping=aes(x=`Term 1`, y=`Term 2`, fill= `Coefficient.nb`)) +
    scale_fill_gradient2(limits = c(lim.2.low,lim.2.high),low = brewer.pal(8,"Set1")[2],
                         mid = "white",
                         high = brewer.pal(8,"Set1")[1],
                         midpoint = 0,
                         guide = "colorbar") +
    
    labs(fill="Second \nOrder ")+
    new_scale_fill()+
    geom_tile(data = subset(epistasis.biochem.df2,part %in% c("D")),
              mapping=aes(`Term 1`, `Term 2`, fill= `Coefficient.nb`)) +
    scale_fill_gradient2(limits= c(lim.higher.low,lim.higher.high),
                         low ="#998ec3",
                         mid = "white",
                         high = "#f1a340",
                         midpoint = 0,
                         na.value = "#303030",
                         guide = "colorbar") +
    labs(fill="Higher \nOrder ")+
    xlab("Mutation 1") +
    ylab("Mutation 2") +
    theme_classic()+
    facet_grid(rows=vars(second),scales="free",space="free_y")+
    #facet_grid(second~Antibody,scales="free",space="free_y")+
    
    #facet_wrap( ~ Antibody, ncol = 2)+
    
    theme(axis.text.x = element_text(angle = 90,hjust = 1, vjust = 0.5),
          axis.text=element_text(size=9.5,color="black"),
          legend.text=element_text(size=7,hjust = 0.5, vjust = 0.5,angle=0),
          legend.title=element_text(size=10,hjust = 0.5, vjust = 1),
          axis.title=element_text(size=13),
          legend.position="right",
          legend.key.size = unit(0.6, "lines"),
          strip.text = element_text(size = 8),
          strip.switch.pad.grid = unit(0.1,"cm"),
          legend.spacing.x = unit(0, 'cm'))
  ggsave(paste0("figures/Figure2_",i,"_epistasis_terms.png"), width = 4, height = 3.2, units = "in")
  ggsave(paste0("figures/THOMAS_IGNORE_Figure2_",i,"_epistasis_terms.png"), width = 4, height = 3.2, units = "in")
}
```

Figure 3B; Epistasis model predictions
```{r epistasis predictions}
Kds_simplified$Log10Kd.epi <- ifelse(Kds_simplified$Log10Kd.nb < 6 | is.na(Kds_simplified$Log10Kd.nb),5,
                                     Kds_simplified$Log10Kd.nb)
#antibody.considered <- "CoV555"
#Kds_simplified.justone <- subset(Kds_simplified,Antibody==antibody.considered)

for(i in c(antibodies,"ACE2")){
  one.prediction <- read.table(file = paste0('data_lasso/lasso_prediction_',i,'_aic_opt_order_biochem.txt'),
                               sep = '\t', 
                               header = TRUE,
                               col.names = c("Genotype","Measured","Predicted.Kd","Predicted.Kd.Linear"),
                               colClasses=c("character","numeric","numeric","numeric"))
  one.prediction$Antibody <- i
  assign(paste0("one.prediction.", i),one.prediction)
}
one.prediction.all <- rbind(one.prediction.CB6,one.prediction.CoV555,one.prediction.REGN10987,one.prediction.S309,one.prediction.ACE2)
one.prediction.all$Antibody <- factor(one.prediction.all$Antibody, levels=c(antibodies,"ACE2"))
Kds_simplified$Predicted <-  one.prediction.all$Predicted.Kd[
  match(interaction(Kds_simplified$Genotype,Kds_simplified$Antibody),
        interaction(one.prediction.all$Genotype,one.prediction.all$Antibody))]

Kds_simplified$Predicted.epi <- ifelse(Kds_simplified$Predicted < 6,
                                       5,
                                       Kds_simplified$Predicted)
#cor(S309.prediction$Measured,S309.prediction$Predicted,use = "na.or.complete")^2
#corr <- cor(Kds_simplified.justone$Log10Kd.epi,Kds_simplified.justone$Predicted.epi,use = "na.or.complete")^2
Kds_simplified$grouping <- case_when(
  (Kds_simplified$Antibody== "CB6" & 
     Kds_simplified$`Mutation 417` == "0") ~ "K417",
  (Kds_simplified$Antibody== "CB6" & 
     Kds_simplified$`Mutation 417` == "1") ~ "N417",
  
  (Kds_simplified$Antibody== "CoV555" & 
     Kds_simplified$`Mutation 493` == "0"&
     Kds_simplified$`Mutation 484` == "0") ~ "Q493/E484",
  (Kds_simplified$Antibody== "CoV555" & 
     Kds_simplified$`Mutation 493` == "1"&
     Kds_simplified$`Mutation 484` == "1") ~ "R493/A484",
  (Kds_simplified$Antibody== "CoV555" & 
     Kds_simplified$`Mutation 493` == "1" &
     Kds_simplified$`Mutation 484` == "0") ~ "R493/E484",
  (Kds_simplified$Antibody== "CoV555" & 
     Kds_simplified$`Mutation 493` == "0" &
     Kds_simplified$`Mutation 484` == "1") ~ "Q493/A484",
  
  (Kds_simplified$Antibody== "REGN10987" & 
     Kds_simplified$`Mutation 440` == "0"&
     Kds_simplified$`Mutation 446` == "0") ~ "N440/G446",
  (Kds_simplified$Antibody== "REGN10987" & 
     Kds_simplified$`Mutation 440` == "1"&
     Kds_simplified$`Mutation 446` == "1") ~ "K440/S446",
  (Kds_simplified$Antibody== "REGN10987" & 
     Kds_simplified$`Mutation 440` == "1" &
     Kds_simplified$`Mutation 446` == "0") ~ "K440/G446",
  (Kds_simplified$Antibody== "REGN10987" & 
     Kds_simplified$`Mutation 440` == "0" &
     Kds_simplified$`Mutation 446` == "1") ~ "N440/S446",
  
  (Kds_simplified$Antibody== "S309") ~ "All S309",
  (Kds_simplified$Antibody== "ACE2") ~ "All ACE2"
  
)

Kds_simplified$grouping <- factor(Kds_simplified$grouping,levels=c("K417","N417",
                                                                   "Q493/E484","R493/E484","Q493/A484","R493/A484",
                                                                   "N440/G446","K440/G446","N440/S446","K440/S446",
                                                                   "All S309", "All ACE2"))
Kds_simplified$Antibody <- factor(Kds_simplified$Antibody,levels=c(antibodies,"ACE2"))

Kds_simplified$Accuracy <- ifelse((Kds_simplified$Log10Kd.epi <= 6 & Kds_simplified$Predicted.epi <= 6)|
                                    (Kds_simplified$Log10Kd.epi > 6 & Kds_simplified$Predicted.epi > 6),TRUE,FALSE )
Kds_prediction_plot <- subset(Kds_simplified,Filtered | Log10Kd.epi < 6.001)
Kds_prediction_plot$grouping <- factor(Kds_prediction_plot$grouping,levels=c("K417","N417",
                                                                             "Q493/E484","R493/E484","Q493/A484","R493/A484",
                                                                             "N440/G446","K440/G446","N440/S446","K440/S446",
                                                                             "All S309","All ACE2"))
CB6.Accuracy <- mean(subset(Kds_prediction_plot,Antibody=="CB6")$Accuracy)
REGN10987.Accuracy <- mean(subset(Kds_prediction_plot,Antibody=="REGN10987")$Accuracy)
S309.Accuracy <- mean(subset(Kds_prediction_plot,Antibody=="S309")$Accuracy)
CoV555.Accuracy <- mean(subset(Kds_prediction_plot,Antibody=="CoV555")$Accuracy)
ACE2.Accuracy <- mean(na.omit(subset(Kds_prediction_plot,Antibody=="ACE2"))$Accuracy)

Kds_prediction_plot2 <- subset(Kds_prediction_plot,Predicted.epi > 6 & Log10Kd.epi > 6)

CB6.r <- cor(subset(Kds_prediction_plot2,Antibody=="CB6" )$Predicted.epi,
             subset(Kds_prediction_plot2,Antibody=="CB6")$Log10Kd.epi)^2
CoV555.r <- cor(subset(Kds_prediction_plot2,Antibody=="CoV555")$Predicted.epi,
                subset(Kds_prediction_plot2,Antibody=="CoV555")$Log10Kd.epi)^2
REGN10987.r <- cor(subset(Kds_prediction_plot2,Antibody=="REGN10987")$Predicted.epi,
                   subset(Kds_prediction_plot2,Antibody=="REGN10987")$Log10Kd.epi)^2
S309.r <- cor(subset(Kds_prediction_plot2,Antibody=="S309")$Predicted.epi,
              subset(Kds_prediction_plot2,Antibody=="S309")$Log10Kd.epi)^2
ACE2.r <- cor(subset(Kds_prediction_plot2,Antibody=="ACE2")$Predicted.epi,
              subset(Kds_prediction_plot2,Antibody=="ACE2")$Log10Kd.epi)^2

legends <- data.frame(Antibody.used = factor(c("LY-CoV016 (4th)","LY-CoV555 (3rd)","REGN10987 (2nd)","S309 (4th)","ACE2 (5th)"),
                                        levels=c("LY-CoV016 (4th)","LY-CoV555 (3rd)","REGN10987 (2nd)","S309 (4th)","ACE2 (5th)")),
                      R2 = c(CB6.r,CoV555.r,REGN10987.r,S309.r,ACE2.r),
                      Acc = c(CB6.Accuracy,CoV555.Accuracy,REGN10987.Accuracy,S309.Accuracy,ACE2.Accuracy))

Kds_prediction_plot$Antibody.used <- ifelse(Kds_prediction_plot$Antibody=="CB6","LY-CoV016 (4th)",
                                             ifelse(Kds_prediction_plot$Antibody=="CoV555","LY-CoV555 (3rd)",
                                                    ifelse(Kds_prediction_plot$Antibody=="REGN10987","REGN10987 (2nd)",
                                                           ifelse(Kds_prediction_plot$Antibody=="S309","S309 (4th)",
                                                                  ifelse(Kds_prediction_plot$Antibody=="ACE2","ACE2 (5th)",NA
                                                           )))))

Kds_prediction_plot$Antibody.used<-factor(Kds_prediction_plot$Antibody.used,
                                           levels = c("LY-CoV016 (4th)","LY-CoV555 (3rd)","REGN10987 (2nd)","S309 (4th)","ACE2 (5th)")
)

ggplot(Kds_prediction_plot, 
       aes(Log10Kd.epi, Predicted.epi)) + 
  facet_wrap( ~ Antibody.used, ncol = 3)+
  geom_jitter(size = 0.5,alpha=0.7,aes(color=grouping)) +
  #scale_color_manual(values=getPalettePairedLong(6)[2:5]) +
  scale_color_manual(values=getPalettePairedLong(13)[2:13]) +
  labs(x=expression(paste("Measured ",(-log~italic(K)[paste("D", ",", "app")]))),
       y=expression(paste("Predicted ",(-log~italic(K)[paste("D", ",", "app")])))) +
  labs(color='Mutation') +
  geom_abline(color="black",linetype="dashed")+
  scale_y_continuous(breaks=c(5,6:11),labels=c("N",6:11))+
  scale_x_continuous(breaks=c(5,6:11),labels=c("N",6:11))+
  #geom_text_repel(data=subset(second.structure,epistasis>0.3 | epistasis<(-0.15)),
  #          aes(distance,epistasis,label=labels),size=3.5,color="black")+
  #guides(color=guide_legend(ncol=2))+
  
  #geom_text_repel(data=subset(second.structure,epistasis>0.3 | epistasis<(-0.15)),
  #          aes(distance,epistasis,label=labels),size=3.5,color="black")+
  theme_classic()+
  geom_richtext(data = legends,
                mapping = aes(x = 5.2, y = 9.8, label = paste0("R<sup>2</sup> = ", round(R2,2),"<br>Acc = ",round(Acc,2))),
                hjust=0,
                size=3.5,
                #label.padding=unit(c(0,0,0,0),"lines"),
                fill = NA,
                label.color = NA)+
  guides(colour = guide_legend(override.aes = list(size=3),ncol=6))+
  theme(axis.text.x = element_text(angle = 0),
        axis.text=element_text(size=12,color="black"),
        legend.position = "none",
        legend.text=element_text(size=10),
        legend.title=element_blank(),
        legend.key.size = unit(0.5, "lines"),
        axis.title.x=element_text(size=13),
        axis.title.y=element_text(size=12.7),
        legend.spacing = unit(4, "pt"), 
        legend.key.height = unit(8, "pt"), 
        legend.spacing.x = unit(4, "pt"), 
        legend.key.width  = unit(4, "pt"), 
        legend.margin=margin(4,4,4,4))
ggsave("figures/Figure3_prediction_by_epistasis_legend_two_optimal_order.png", width = 5.2, height = 3.5, units = "in")


```
## Figure 4 - Comparison with ACE2 Data
Figure 4C: Phenotypicradeoffs
```{r tradeoffs}
Kds.short.tradeoff[paste0("Mutation_",nn)] <- Kds_simplified[1:32768,paste0("Mutation ",nn)]
Kds.short.tradeoff$Mean_Ab <- rowMeans(Kds.short.tradeoff[,3:6])
Kds.short.tradeoff$Mean_Ab <- rowMeans(Kds.short.tradeoff[,3:6])
Kds.short.tradeoff$Max_Ab <- apply(Kds.short.tradeoff[,3:6],1,max)
Kds.short.tradeoff$Max_Ab_which <- ifelse(Kds.short.tradeoff$Max_Ab==Kds.short.tradeoff$CB6,
                                          "CB6",
                                          ifelse(Kds.short.tradeoff$Max_Ab==Kds.short.tradeoff$CoV555,
                                                 "CoV555",
                                                 ifelse(Kds.short.tradeoff$Max_Ab==Kds.short.tradeoff$REGN10987,
                                                        "REGN10987",
                                                        "S309"
                                                 )))
Kds.short.tradeoff$Max_Ab_which <- factor(Kds.short.tradeoff$Max_Ab_which,levels=antibodies)
corr1 <- cor(Kds.short.tradeoff$Max_Ab,Kds.short.tradeoff$ACE2)^2
corr2 <- cor(subset(Kds.short.tradeoff,Mutation_501=="0"&Mutation_498=="0")$Max_Ab,
             subset(Kds.short.tradeoff,Mutation_501=="0"&Mutation_498=="0")$ACE2)^2

Kds.short.tradeoff$Mut498_501 <- interaction(
               Kds.short.tradeoff$`Mutation_498`,Kds.short.tradeoff$`Mutation_501`,sep="",lex.order=TRUE)

Kds.short.tradeoff2 <- Kds.short.tradeoff
Kds.short.tradeoff2$ACE2 <- ifelse(Kds.short.tradeoff2$ACE2 < 6, NA,Kds.short.tradeoff2$ACE2)
Kds.short.tradeoff2$CB6 <- ifelse(Kds.short.tradeoff2$CB6 < 6, NA,Kds.short.tradeoff2$CB6)
Kds.short.tradeoff2$CoV555 <- ifelse(Kds.short.tradeoff2$CoV555 < 6, NA,Kds.short.tradeoff2$CoV555)
Kds.short.tradeoff2$REGN10987 <- ifelse(Kds.short.tradeoff2$REGN10987 < 6, NA,Kds.short.tradeoff2$REGN10987)
Kds.short.tradeoff2$S309 <- ifelse(Kds.short.tradeoff2$S309 < 6, NA,Kds.short.tradeoff2$S309)

for(i in 1:4){
  for(j in 5:5){
    for(k in 1:4){
    ab1 <- c(antibodies,"ACE2")[i]
    ab2 <- c(antibodies,"ACE2")[j]
    sab1 <- ifelse(ab1=="CB6","LY-CoV016",
                   ifelse(ab1=="CoV555","LY-CoV555",ab1))
    mut <- levels(Kds.short.tradeoff$Mut498_501)[k]
    Kds.tradeoff.considered <- subset(Kds.short.tradeoff,Mut498_501==mut)
    #Kds.tradeoff.considered <- subset(Kds.tradeoff.considered,ab1>5)
    print(round(cor(Kds.tradeoff.considered[,ab1],Kds.tradeoff.considered[,ab2], use="complete.obs"),2))
    }
  }
}

for(i in 1:4){
  for(j in 5:5){
    ab1 <- c(antibodies,"ACE2")[i]
    ab2 <- c(antibodies,"ACE2")[j]
    sab1 <- ifelse(ab1=="CB6","LY-CoV016",
                   ifelse(ab1=="CoV555","LY-CoV555",ab1))
    correlation00 <- round(cor(subset(Kds.short.tradeoff2,Mut498_501=="00")[,ab1],
                             subset(Kds.short.tradeoff2,Mut498_501=="00")[,ab2], use="complete.obs"),2)
    correlation11 <- round(cor(subset(Kds.short.tradeoff2,Mut498_501=="11")[,ab1],
                             subset(Kds.short.tradeoff2,Mut498_501=="11")[,ab2], use="complete.obs"),2)
    Kds.template <- Kds.short.tradeoff
    Kds.template$Ab1 <- Kds.short.tradeoff[,ab1]
    Kds.template$Ab2 <- Kds.short.tradeoff[,ab2]
    ggplot(Kds.template,aes(x=Ab1,y=Ab2))+
      #ggplot(subset(Kds.short.tradeoff,Mutation_501=="0"&Mutation_498=="0"),aes(x=ACE2,y=Max_Ab,color=Max_Ab_which))+
      #geom_line(aes(color= `Mutation`),size=1.25,alpha=0.8)+
      #geom_errorbar(aes(ymin=`KD Effect`-STE, ymax=`KD Effect`+STE),color="black",width=0.0003,
      #              colour = "#969696",size=0.5,alpha=0.5)+
      geom_point(size = 0.5 ,alpha=0.5,aes(color=interaction(
               `Mutation_498`,`Mutation_501`,sep="/",lex.order=TRUE)))+
      scale_color_manual(values=getPalettePairedLong(6)[2:5]) +
      scale_x_continuous(breaks=c(5,6:11),labels=c("NB",6:11))+
      scale_y_continuous(breaks=c(5,6:11),labels=c("NB",6:11))+
      #geom_smooth(method='lm',color="black")+
      geom_richtext(data = legends,
                    mapping = aes(x = 5, 
                                  y = 10.2, 
                                  label = paste0("r = ",correlation00)
                                  ),
                    hjust=0,
                    color = getPalettePairedLong(6)[2],
                    size=3.5,
                    #label.padding=unit(c(0,0,0,0),"lines"),
                    fill = NA,
                    label.color = NA)+
      geom_richtext(data = legends,
                    mapping = aes(x = 5, 
                                  y = 9.8, 
                                  label = paste0("r = ",correlation11)
                                  ),
                    hjust=0,
                    color = getPalettePairedLong(6)[5],
                    size=3.5,
                    #label.padding=unit(c(0,0,0,0),"lines"),
                    fill = NA,
                    label.color = NA)+
      #scale_color_manual(values = getPalettePairedLong(15))+
      
      labs(x=bquote(.(sab1)~(-log~italic(K)[paste("D", ",", "app")])),
           y=bquote(.(ab2)~(-log~italic(K)[paste("D", ",", "app")])),
           color = "Antibody") +
      #geom_text_repel(data=subset(first.structure,buriedSA>0),
      #          aes(buriedSA-3.5,ACE2_Effect+.1,label=Term),size=4,color="black")+
      theme_classic()+
      #xlim(6,10.5)+
      #xlim(0,130)+
      #geom_hline(yintercept = 0 , colour = "black", linetype = "dashed")+
      guides(colour = guide_legend(override.aes = list(size=3)))+
      theme(
            axis.text=element_text(size=12,color="black"),
            axis.text.x = element_text(size=10,angle = 0),
            axis.text.y = element_text(size=10),
            legend.position = "none",
            legend.text=element_text(size=10),
            #legend.title=element_blank(),
            legend.key.size = unit(0.5, "lines"),
            axis.title.y=element_text(size=11),
            axis.title.x=element_text(size=10.5),
            legend.spacing = unit(4, "pt"), 
            legend.key.height = unit(8, "pt"), 
            legend.spacing.x = unit(4, "pt"), 
            legend.key.width  = unit(4, "pt"), 
            legend.margin=margin(4,4,4,4),
            legend.title = element_text(size=11))
    ggsave(paste0("figures/FigureSI_dot_plots_tradeoff_",ab1,"_",ab2,".png"),
           width = 2.2, height = 2.2, units = "in")
  }
}

for(i in 1:3){
  for(j in (i+1):4){
    ab1 <- c(antibodies,"ACE2")[i]
    ab2 <- c(antibodies,"ACE2")[j]
    print(round(cor(Kds.short.tradeoff[,ab1],Kds.short.tradeoff[,ab2], use="complete.obs"),2))
  }
}

for(i in 1:3){
  for(j in (i+1):4){
    ab1 <- c(antibodies,"ACE2")[i]
    ab2 <- c(antibodies,"ACE2")[j]
    correlation <- round(cor(Kds.short.tradeoff[,ab1],Kds.short.tradeoff[,ab2], use="complete.obs")^2,2)
    Kds.template <- Kds.short.tradeoff
    Kds.template$Ab1 <- Kds.short.tradeoff[,ab1]
    Kds.template$Ab2 <- Kds.short.tradeoff[,ab2]
    ggplot(Kds.template,aes(x=Ab1,y=Ab2))+
      #ggplot(subset(Kds.short.tradeoff,Mutation_501=="0"&Mutation_498=="0"),aes(x=ACE2,y=Max_Ab,color=Max_Ab_which))+
      #geom_line(aes(color= `Mutation`),size=1.25,alpha=0.8)+
      
      +
      #geom_errorbar(aes(ymin=`KD Effect`-STE, ymax=`KD Effect`+STE),color="black",width=0.0003,
      #              colour = "#969696",size=0.5,alpha=0.5)+
      geom_point(size = 0.5 ,alpha=0.5,color="black")+
      scale_color_manual(values=getPalettePairedLong(6)[2:5]) +
      geom_smooth(method='lm',color="black")+
      geom_richtext(data = legends,
                    mapping = aes(x = 6.1, 
                                  y = 10.1, 
                                  label = paste0("R<sup>2</sup> = ",correlation)
                                  ),
                    hjust=0,
                    size=3.5,
                    #label.padding=unit(c(0,0,0,0),"lines"),
                    fill = NA,
                    label.color = NA)+
      #scale_color_manual(values = getPalettePairedLong(15))+
      labs(x=expression(paste(ab1,(-log~italic(K)[paste("D", ",", "app")]))),
           y=expression(paste(ab2,(-log~italic(K)[paste("D", ",", "app")]))),
           color = "Antibody") +
      #geom_text_repel(data=subset(first.structure,buriedSA>0),
      #          aes(buriedSA-3.5,ACE2_Effect+.1,label=Term),size=4,color="black")+
      theme_classic()+
      #xlim(6,10.5)+
      #xlim(0,130)+
      #geom_hline(yintercept = 0 , colour = "black", linetype = "dashed")+
      guides(colour = guide_legend(override.aes = list(size=3)))+
      theme(
            axis.text=element_text(size=12,color="black"),
            axis.text.x = element_text(size=14,angle = -1),
            axis.text.y = element_text(size=12),
            legend.position = "none",
            legend.text=element_text(size=10,vjust=1),
            #legend.title=element_blank(),
            legend.key.size = unit(0.5, "lines"),
            axis.title=element_text(size=14),
            legend.spacing = unit(4, "pt"), 
            legend.key.height = unit(8, "pt"), 
            legend.spacing.x = unit(4, "pt"), 
            legend.key.width  = unit(4, "pt"), 
            legend.margin=margin(4,4,4,4),
            legend.title = element_text(size=11))
    #ggsave(paste0("figures/FigureSI_dot_plots_tradeoff_",ab1,"_",ab2,".png"),width = 2.2, height = 2.2, units = "in")
  }
}


ggplot(Kds.short.tradeoff,aes(x=S309,y=ACE2))+
#ggplot(subset(Kds.short.tradeoff,Mutation_501=="0"&Mutation_498=="0"),aes(x=ACE2,y=Max_Ab,color=Max_Ab_which))+
  #geom_line(aes(color= `Mutation`),size=1.25,alpha=0.8)+
  geom_point(size = 0.5 ,alpha=0.5,aes(color=interaction(
    `Mutation_498`,`Mutation_501`,sep="/",lex.order=TRUE))) +
  #geom_errorbar(aes(ymin=`KD Effect`-STE, ymax=`KD Effect`+STE),color="black",width=0.0003,
  #              colour = "#969696",size=0.5,alpha=0.5)+
  scale_color_manual(values=getPalettePairedLong(6)[2:5]) +
  geom_smooth(method='lm',color="black")+
  geom_richtext(data = legends,
                mapping = aes(x = 6.1, 
                              y = 10.6, 
                              label = paste0("R<sup>2</sup> = ",
                                             round(cor(Kds.short.tradeoff$S309,
                                                       Kds.short.tradeoff$ACE2)^2,2))
                              ),
                hjust=0,
                size=3.5,
                #label.padding=unit(c(0,0,0,0),"lines"),
                fill = NA,
                label.color = NA)+
  #scale_color_manual(values = getPalettePairedLong(15))+
  labs(x=expression(paste("ACE2 ",(-log~italic(K)[paste("D", ",", "app")]))),
       y=expression(paste("Max mAb ",(-log~italic(K)[paste("D", ",", "app")]))),
       color = "Antibody") +
  #geom_text_repel(data=subset(first.structure,buriedSA>0),
  #          aes(buriedSA-3.5,ACE2_Effect+.1,label=Term),size=4,color="black")+
  theme_classic()+
  xlim(6,10.5)+
  #xlim(0,130)+
  #geom_hline(yintercept = 0 , colour = "black", linetype = "dashed")+
  guides(colour = guide_legend(override.aes = list(size=3)))+
  theme(
        axis.text=element_text(size=12,color="black"),
        axis.text.x = element_text(size=14,angle = 0),
        axis.text.y = element_text(size=12),
        legend.position = "right",
        legend.text=element_text(size=10),
        #legend.title=element_blank(),
        legend.key.size = unit(0.5, "lines"),
        axis.title=element_text(size=14),
        legend.spacing = unit(4, "pt"), 
        legend.key.height = unit(8, "pt"), 
        legend.spacing.x = unit(4, "pt"), 
        legend.key.width  = unit(4, "pt"), 
        legend.margin=margin(4,4,4,4),
        legend.title = element_text(size=11))


ggplot(Kds.template,aes(x=Ab1,y=Ab2))+
    #ggplot(subset(Kds.short.tradeoff,Mutation_501=="0"&Mutation_498=="0"),aes(x=ACE2,y=Max_Ab,color=Max_Ab_which))+
    #geom_line(aes(color= `Mutation`),size=1.25,alpha=0.8)+
    #geom_errorbar(aes(ymin=`KD Effect`-STE, ymax=`KD Effect`+STE),color="black",width=0.0003,
    #              colour = "#969696",size=0.5,alpha=0.5)+
    geom_point(size = 0.5 ,alpha=0.5,aes(color=interaction(
             `Mutation_498`,`Mutation_501`,sep="/",lex.order=TRUE)))+
    scale_color_manual(labels=c("0/0" = "Q498/N501", 
                            "1/0" = "R498/N501",
                            "0/1" = "Q498/Y501",
                            "1/1" = "R498/Y501"),
                       values=getPalettePairedLong(6)[2:5]) +
    scale_x_continuous(breaks=c(5,6:11),labels=c("NB",6:11))+
    scale_y_continuous(breaks=c(5,6:11),labels=c("NB",6:11))+
    geom_smooth(method='lm',color="black")+
    geom_richtext(data = legends,
                  mapping = aes(x = 5.2, 
                                y = 10.2, 
                                label = paste0("r = ",correlation)
                                ),
                  hjust=0,
                  size=3.5,
                  #label.padding=unit(c(0,0,0,0),"lines"),
                  fill = NA,
                  label.color = NA)+
    #scale_color_manual(values = getPalettePairedLong(15))+
    
    labs(x=bquote(.(sab1)~-log~italic(K)[paste("D", ",", "app")]),
         y=bquote(.(ab2)~-log~italic(K)[paste("D", ",", "app")]),
         color = "") +
    #geom_text_repel(data=subset(first.structure,buriedSA>0),
    #          aes(buriedSA-3.5,ACE2_Effect+.1,label=Term),size=4,color="black")+
    theme_classic()+
    #xlim(6,10.5)+
    #xlim(0,130)+
    #geom_hline(yintercept = 0 , colour = "black", linetype = "dashed")+
    guides(colour = guide_legend(override.aes = list(size=3),ncol=1,byrow=TRUE))+
    theme(
          axis.text=element_text(size=12,color="black"),
          axis.text.x = element_text(size=10,angle = 0),
          axis.text.y = element_text(size=10),
          legend.position = "bottom",
          legend.text=element_text(size=12,vjust=0.4),
          #legend.title=element_blank(),
          legend.key.size = unit(1, "lines"),
          axis.title=element_text(size=11),
          legend.spacing.y = unit(6,"pt"), 
          legend.key.height = unit(4, "pt"), 
          legend.key.width  = unit(4, "pt"), 
          legend.margin=margin(6,6,6,6),
          legend.title = element_text(size=11))
ggsave(paste0("figures/FigureSI_dot_plots_tradeoff_legend.png"),
       width = 8.8, height = 4, units = "in")

```
Figure 4A-B: ACE2 effects
```{r}
Kds.short.tradeoff$Classes.escapes <-  0

Kds.short.tradeoff$`Antibodies Escaped` <- (Kds.short.tradeoff$CB6<=6.00) + 
  (Kds.short.tradeoff$CoV555<=6.00) +
  (Kds.short.tradeoff$REGN10987<=6.000)
Kds.short.tradeoff$`Antibodies Escaped` <- factor(Kds.short.tradeoff$`Antibodies Escaped`,levels=0:3)

compensatory.class <- function(genox){
  ifelse(substr(genox,13,14)=="11",return("with"),
         ifelse(substr(genox,13,14)=="00",return("without"),
                ifelse(substr(genox,13,14)=="10",return(NA),
                       ifelse(substr(genox,13,14)=="01",return(NA),NA))))
  #return((substr(genox,13,14)=="11"))
}

Kds.short.tradeoff$`Compensatory Mutations` <-  sapply(Kds.short.tradeoff$Genotype,compensatory.class)
#Kds.short.tradeoff$`Compensatory Mutations` <- ifelse(Kds.short.tradeoff$`Compensatory Mutations`,'Present','Absent')
Kds.short.tradeoff$`Compensatory Mutations` <- factor(Kds.short.tradeoff$`Compensatory Mutations`,
                                                      levels=c('without','with'))

Kds.short.tradeoff$Class <- str_count(Kds.short.tradeoff$Genotype,"1") #initialize variable "Class", which is #hamming distance from Wuhan
Kds.short.tradeoff$Class <- factor(Kds.short.tradeoff$Class,levels=0:15)

ggplot(na.omit(Kds.short.tradeoff),aes(x=Class,y=ACE2,fill =`Antibodies Escaped`,color=`Compensatory Mutations`))+
  #geom_jitter(size=1.25,alpha=0.8)+
  geom_boxplot(size=0.4,outlier.size = 0.4)+
  #geom_line(aes(color=Classes.escapes,linetype=compensatory),size = 1 ,alpha=0.5) +
  #geom_errorbar(aes(ymin=`KD Effect`-STE, ymax=`KD Effect`+STE),color="black",width=0.0003,
  #              colour = "#969696",size=0.5,alpha=0.5)+
  scale_fill_manual(values = getYellowtoRed(4)[1:4])+
  scale_color_manual(values = c("gray80","gray10"))+
  labs(y=expression(paste("ACE2 ",(-log~italic(K)[paste("D", ",", "app")]))),
       x="Number of Mutations",
       color = "Compensatory\nMutations",
       fill = "Number of mAbs\nEscaped") +
  #geom_text_repel(data=subset(first.structure,buriedSA>0),
  #          aes(buriedSA-3.5,ACE2_Effect+.1,label=Term),size=4,color="black")+
  theme_classic()+
  #xlim(0,130)+
  geom_hline(yintercept = 9.033694 , colour = "gray50", linetype = "dashed")+
  #guides(color=guide_legend(ncol=2))+
  theme(
    axis.text=element_text(size=10,color="black"),
    axis.text.x = element_text(size=12,angle = 0),
    axis.text.y = element_text(size=12),
    legend.position = "right",
    axis.title=element_text(size=14),
    legend.text = element_text(size=10),
    legend.title = element_text(size=11))
ggsave("figures/FigureSI4_ACE2_across_numb_mutations_with_comp.png", width = 7.5, height = 3, units = "in")

ggplot(na.omit(Kds.short.tradeoff),aes(x=Class,y=ACE2,fill =`Antibodies Escaped`))+
  #geom_jitter(size=1.25,alpha=0.8)+
  geom_boxplot(size=0.4,outlier.size = 0.4)+
  #geom_line(aes(color=Classes.escapes,linetype=compensatory),size = 1 ,alpha=0.5) +
  #geom_errorbar(aes(ymin=`KD Effect`-STE, ymax=`KD Effect`+STE),color="black",width=0.0003,
  #              colour = "#969696",size=0.5,alpha=0.5)+
  scale_fill_manual(values = getYellowtoRed(4)[1:4])+
  #scale_color_manual(values = c("gray80","gray10"))+
  labs(y=expression(paste("ACE2 ",(-log~italic(K)[paste("D", ",", "app")]))),
       x="Number of Mutations",
       color = "Compensatory\nMutations",
       fill = "Number of mAbs\nEscaped") +
  #geom_text_repel(data=subset(first.structure,buriedSA>0),
  #          aes(buriedSA-3.5,ACE2_Effect+.1,label=Term),size=4,color="black")+
  theme_classic()+
  #xlim(0,130)+
  geom_hline(yintercept = 9.033694 , colour = "gray50", linetype = "dashed")+
  #guides(color=guide_legend(ncol=2))+
  theme(
    axis.text=element_text(size=10,color="black"),
    axis.text.x = element_text(size=12,angle = 0),
    axis.text.y = element_text(size=12),
    legend.position = "right",
    axis.title=element_text(size=14),
    legend.text = element_text(size=10),
    legend.title = element_text(size=11))
ggsave("figures/Figure4_ACE2_across_numb_mutations_without_comp.png", width = 8.2, height = 2.5, units = "in")
###per abs
Kds.short.tradeoff$`CB6 Escaped` <-  factor(ifelse(Kds.short.tradeoff$CB6 <= 6.00,'Yes','No'),
                                            levels=c('Yes','No'))

Kds.short.tradeoff$`CoV555 Escaped` <-  factor(ifelse(Kds.short.tradeoff$CoV555 <= 6.00,'Yes','No'),
                                            levels=c('Yes','No'))

Kds.short.tradeoff$`REGN10987 Escaped` <-  factor(ifelse(Kds.short.tradeoff$REGN10987 <= 6.00,'Yes','No'),
                                               levels=c('Yes','No'))


Kds.short.tradeoff$`Antibodies` <- factor(paste0(as.numeric(Kds.short.tradeoff$CB6 <= 6.00),
                                                         as.numeric(Kds.short.tradeoff$CoV555 <= 6.00),
                                                         as.numeric(Kds.short.tradeoff$REGN10987 <= 6.00)),
                                                  levels = c('000','100','010','001','110','101','011','111'))


mean(subset(Kds.short.tradeoff,Antibodies=='000' & `Compensatory Mutations`=="without")$ACE2)
ggplot(na.omit(Kds.short.tradeoff),aes(x=Antibodies,y=ACE2,fill =`Compensatory Mutations`))+
  #geom_jitter(size=1.25,alpha=0.8)+
  #geom_jitter(aes(color=Class),alpha=0.2)+
 # geom_violin(alpha=0.8,trim=FALSE)+
  
  geom_boxplot( color="black", alpha = 0.8) +
  #geom_boxplot()+
  #geom_line(aes(color=Classes.escapes,linetype=compensatory),size = 1 ,alpha=0.5) +
  #geom_errorbar(aes(ymin=`KD Effect`-STE, ymax=`KD Effect`+STE),color="black",width=0.0003,
  #              colour = "#969696",size=0.5,alpha=0.5)+
  scale_fill_manual(values = c("gray90","gray10"))+
  #scale_color_manual(values = getYellowtoRed(16))+
  labs(y=expression(atop("ACE2",(-log~italic(K)[paste("D", ",", "app")]))),
       x="Antibodies Escaped",
       
       fill="Q498R/N501Y") +
  scale_x_discrete(labels=c("000" = "None", 
                            "100" = "LY-CoV016",
                            "010" = "LY-CoV555",
                            "001" = "REGN10987",
                            "110" = "LY-CoV016,\nLY-CoV555",
                            "101" = "LY-CoV016,\nREGN10987",
                            "011" = "LY-CoV555,\nREGN10987",
                            "111" = "All Three"),
                   expand=c(0.1, 0),drop=TRUE)+
  #geom_text_repel(data=subset(first.structure,buriedSA>0),
  #          aes(buriedSA-3.5,ACE2_Effect+.1,label=Term),size=4,color="black")+
  theme_classic()+
  #xlim(0,130)+
  geom_hline(yintercept = 9.033694 , colour = "black", linetype = "dotted")+
  geom_hline(yintercept = mean(subset(Kds.short.tradeoff,Antibodies=='000' & `Compensatory Mutations`=="without")$ACE2),
             colour = "#121212", linetype = "dashed")+
  
  #guides(color=guide_legend(ncol=2))+
  theme(
    axis.text=element_text(size=10,color="black"),
    axis.text.x = element_text(size=10,angle = 90,vjust=0.5,hjust=1),
    axis.text.y = element_text(size=14),
    legend.position = "right",
    axis.title.x=element_text(size=14),
    axis.title.y=element_text(size=12.5),
    legend.text = element_text(size=10),
    legend.title = element_text(size=11))
ggsave("figures/Figure4_ACE2_across_abs_escaped.png", width = 8.2, height = 2.5, units = "in")
```

```{r}
phylogeny_stuff <- read.delim("data/fig4_phylogeny.csv" ,header = TRUE, sep = ",", dec = ".")
phylogeny_stuff$enrichment <- (phylogeny_stuff$philogeny_501498/phylogeny_stuff$philogeny_all)*13/15
phylogeny_stuff.long <- reshape(phylogeny_stuff, 
                                direction = "long",
                                varying = list(names(phylogeny_stuff)[2:9]),
                                v.names = "Value",
                                idvar = "Mutation",
                                timevar = "Antibody",
                                times = c(names(phylogeny_stuff)[2:6],"Freq.","Freq. with\nQ498R, N501Y", "Enrichment with\nQ498R, N501Y"))
phylogeny_stuff.long$Antibody <- factor(phylogeny_stuff.long$Antibody,
                                        levels=c(names(phylogeny_stuff)[2:6],"Freq. with\nQ498R, N501Y","Freq.","Enrichment with\nQ498R, N501Y"))

phylogeny_stuff.long$Type <- factor(c(rep("Mean Effect",15*5),rep("Frequency",15*2),rep("En",15)),
                                    levels=c("Mean Effect", "Frequency","En"))

phylogeny_stuff.long$Mutation <- factor(phylogeny_stuff.long$Mutation, levels=nn.muts)
phylogeny_stuff.long$Antibody.used <- ifelse(phylogeny_stuff.long$Antibody=="CB6","LY-CoV016",
                                             ifelse(phylogeny_stuff.long$Antibody=="CoV555",
                                                    "LY-CoV555",
                                                    as.character(phylogeny_stuff.long$Antibody)))
  
phylogeny_stuff.long$Antibody.used <- factor(phylogeny_stuff.long$Antibody.used,
                                             levels = c("S309",
                                                        "REGN10987",
                                                        "LY-CoV555",
                                                        "LY-CoV016",
                                                        "ACE2","Freq. with\nQ498R, N501Y","Freq.","Enrichment with\nQ498R, N501Y")
)
ggplot() + 
  geom_tile(data = subset(phylogeny_stuff.long,Type =="Mean Effect"),
            mapping=aes(x=`Mutation`, y=`Antibody.used`, fill= `Value`)) +
  scale_fill_gradient2(low = brewer.pal(8,"Set1")[2],
                       mid = "white",
                       high = brewer.pal(8,"Set1")[1],
                       midpoint = 0,
                       guide = guide_colorbar(order = 1)) +
  #labs(fill="Second \nOrder ")+
  new_scale_fill()+
  geom_tile(data = subset(phylogeny_stuff.long,Type =="En"),
             mapping=aes(x=`Mutation`, y=`Antibody.used`, fill= `Value`)) +
  scale_fill_gradient2(
                       low ="#66a182",
                       mid = "white",
                       high = "#edae49",
                       midpoint = 1,
                       na.value = "#303030",
                       guide = guide_colorbar(order = 2)) +
  #labs(fill="Higher \nOrder ")+
  xlab("Mutation") +
  ylab("") +
  theme_classic()+
  facet_grid(rows=vars(Type),scales="free",space="free_y")+
  #facet_grid(second~Antibody,scales="free",space="free_y")+
  
  #facet_wrap( ~ Antibody, ncol = 2)+
  #coord_equal()+
  theme(axis.text.x = element_text(angle = 90,hjust = 1, vjust = 0.5),
        axis.text=element_text(size=12,color="black"),
        legend.text=element_text(size=10,hjust = 0.5, vjust = 0.5,angle=0),
        legend.title=element_blank(),
        axis.title=element_text(size=14),
        legend.position="right",
        legend.key.size = unit(0.7, "lines"),
        strip.text = element_text(size=8),
        strip.switch.pad.grid = unit(0.1,"cm"),
        legend.spacing.x = unit(0, 'cm'))


ggsave("figures/Figure4_phylogeny.png", width = 6.5, height = 2.7, units = "in")
```